// @emailcheck/email-validator-js/dist/index.js
'use strict';

var psl = require('psl');
var tinyLru = require('tiny-lru');
var stringSimilarityJs = require('string-similarity-js');
var node_dns = require('node:dns');
var net = require('node:net');
var tls = require('node:tls');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var net__namespace = /*#__PURE__*/_interopNamespaceDefault(net);
var tls__namespace = /*#__PURE__*/_interopNamespaceDefault(tls);

class LRUAdapter {
  constructor(maxSize = 1e3, ttlMs = 36e5) {
    this.lru = tinyLru.lru(maxSize, ttlMs);
  }
  get(key) {
    const value = this.lru.get(key);
    return value === void 0 ? null : value;
  }
  async set(key, value, ttlMs) {
    if (ttlMs !== void 0) {
      this.lru.set(key, value);
    } else {
      this.lru.set(key, value);
    }
  }
  async delete(key) {
    this.lru.delete(key);
    return true;
  }
  async has(key) {
    return this.lru.has(key);
  }
  async clear() {
    this.lru.clear();
  }
  size() {
    return this.lru.size;
  }
  /**
   * Get the underlying LRU instance for advanced operations
   */
  getLRU() {
    return this.lru;
  }
}

const DEFAULT_CACHE_OPTIONS = {
  ttl: {
    mx: 36e5,
    disposable: 864e5,
    free: 864e5,
    domainValid: 864e5,
    smtp: 18e5,
    smtpPort: 864e5,
    domainSuggestion: 864e5,
    whois: 36e5
  },
  maxSize: {
    mx: 1e4,
    disposable: 1e4,
    free: 1e4,
    domainValid: 1e4,
    smtp: 1e4,
    smtpPort: 1e4,
    domainSuggestion: 1e4,
    whois: 1e4
  }
};
let defaultCacheInstance = null;
function getDefaultCache() {
  if (!defaultCacheInstance) {
    defaultCacheInstance = {
      mx: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.mx, DEFAULT_CACHE_OPTIONS.ttl.mx),
      disposable: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.disposable, DEFAULT_CACHE_OPTIONS.ttl.disposable),
      free: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.free, DEFAULT_CACHE_OPTIONS.ttl.free),
      domainValid: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.domainValid, DEFAULT_CACHE_OPTIONS.ttl.domainValid),
      smtp: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.smtp, DEFAULT_CACHE_OPTIONS.ttl.smtp),
      smtpPort: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.smtpPort, DEFAULT_CACHE_OPTIONS.ttl.smtpPort),
      domainSuggestion: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.domainSuggestion, DEFAULT_CACHE_OPTIONS.ttl.domainSuggestion),
      whois: new LRUAdapter(DEFAULT_CACHE_OPTIONS.maxSize.whois, DEFAULT_CACHE_OPTIONS.ttl.whois)
    };
  }
  return defaultCacheInstance;
}
function getCacheStore(cache, key) {
  return (cache === null || cache === void 0 ? void 0 : cache[key]) || getDefaultCache()[key];
}
function clearDefaultCache() {
  if (defaultCacheInstance) {
    defaultCacheInstance.mx.clear();
    defaultCacheInstance.disposable.clear();
    defaultCacheInstance.free.clear();
    defaultCacheInstance.domainValid.clear();
    defaultCacheInstance.smtp.clear();
    defaultCacheInstance.smtpPort.clear();
    defaultCacheInstance.domainSuggestion.clear();
    defaultCacheInstance.whois.clear();
  }
}
function resetDefaultCache() {
  defaultCacheInstance = null;
}

const commonEmailDomains = [
  // Popular free email providers
  "gmail.com",
  "yahoo.com",
  "outlook.com",
  "hotmail.com",
  "icloud.com",
  "aol.com",
  "mail.com",
  "protonmail.com",
  "proton.me",
  "yandex.com",
  // Business/Professional email providers
  "google.com",
  "microsoft.com",
  "apple.com",
  "amazon.com",
  "amazonaws.com",
  // Secure/Privacy-focused providers
  "tutanota.com",
  "tuta.com",
  "fastmail.com",
  "posteo.de",
  "mailbox.org",
  "runbox.com",
  "startmail.com",
  // European providers
  "gmx.com",
  "gmx.de",
  "gmx.net",
  "web.de",
  "freenet.de",
  "t-online.de",
  "arcor.de",
  // Business email hosting
  "zoho.com",
  "titan.email",
  "neo.space",
  "rackspace.com",
  "ionos.com",
  "hostinger.com",
  "bluehost.com",
  "siteground.com",
  "dreamhost.com",
  "inmotionhosting.com",
  "godaddy.com",
  "namecheap.com",
  "migadu.com",
  "purelymail.com",
  "infomaniak.com",
  "uberspace.de",
  "manitu.de",
  "hetzner.com",
  "scalahosting.com",
  "hostpapa.com",
  "liquidweb.com",
  "icewarp.com",
  "hostgator.com",
  "a2hosting.com",
  "spacemail.com",
  "mxroute.com",
  // Regional providers
  "qq.com",
  "163.com",
  "126.com",
  "sina.com",
  "naver.com",
  "daum.net",
  "kakao.com"
];
const TYPO_PATTERNS = {
  "gmail.com": ["gmai.com", "gmial.com", "gmali.com", "gmil.com", "gmaill.com", "gmail.co", "gmail.cm"],
  "yahoo.com": ["yaho.com", "yahooo.com", "yahoo.co", "yahoo.cm", "yhaoo.com"],
  "hotmail.com": ["hotmai.com", "hotmial.com", "hotmal.com", "hotmil.com", "hotmail.co", "hotmail.cm"],
  "outlook.com": ["outlok.com", "outloo.com", "outlook.co", "outlook.cm", "outlokk.com"]
};
function getSimilarityThreshold(domain) {
  const length = domain.length;
  if (length <= 6)
    return 0.85;
  if (length <= 10)
    return 0.8;
  return 0.75;
}
function defaultDomainSuggestionMethod(domain, commonDomains) {
  const domainsToCheck = commonDomains || commonEmailDomains;
  const lowerDomain = domain.toLowerCase();
  if (domainsToCheck.includes(lowerDomain)) {
    return null;
  }
  for (const [correctDomain, typos] of Object.entries(TYPO_PATTERNS)) {
    if (typos.includes(lowerDomain)) {
      return {
        original: domain,
        suggested: correctDomain,
        confidence: 0.95
        // High confidence for known typo patterns
      };
    }
  }
  let bestMatch = null;
  const threshold = getSimilarityThreshold(lowerDomain);
  for (const commonDomain of domainsToCheck) {
    const similarity = stringSimilarityJs.stringSimilarity(lowerDomain, commonDomain.toLowerCase());
    if (similarity >= threshold) {
      if (!bestMatch || similarity > bestMatch.similarity) {
        bestMatch = { domain: commonDomain, similarity };
      }
    }
  }
  if (!bestMatch) {
    for (const commonDomain of domainsToCheck) {
      if (Math.abs(lowerDomain.length - commonDomain.length) <= 2) {
        const similarity = stringSimilarityJs.stringSimilarity(lowerDomain, commonDomain.toLowerCase());
        if (similarity >= 0.7) {
          if (!bestMatch || similarity > bestMatch.similarity) {
            bestMatch = { domain: commonDomain, similarity };
          }
        }
      }
    }
  }
  if (bestMatch) {
    if (bestMatch.domain.charAt(0) !== lowerDomain.charAt(0) && bestMatch.similarity < 0.9) {
      return null;
    }
    return {
      original: domain,
      suggested: bestMatch.domain,
      confidence: bestMatch.similarity
    };
  }
  return null;
}
async function defaultDomainSuggestionMethodAsync(domain, commonDomains, cache) {
  return defaultDomainSuggestionMethodImpl(domain, commonDomains, cache);
}
async function defaultDomainSuggestionMethodImpl(domain, commonDomains, cache) {
  if (!domain || domain.length < 3) {
    return null;
  }
  const domainsToCheck = commonDomains || commonEmailDomains;
  const lowerDomain = domain.toLowerCase();
  const cacheKey = `${lowerDomain}:${domainsToCheck.length}`;
  const cacheStore = getCacheStore(cache, "domainSuggestion");
  const cached = await cacheStore.get(cacheKey);
  const resolved = cached && typeof cached === "object" && "then" in cached ? await cached : cached;
  if (resolved !== null && resolved !== void 0) {
    return resolved;
  }
  if (domainsToCheck.includes(lowerDomain)) {
    await cacheStore.set(cacheKey, null);
    return null;
  }
  for (const [correctDomain, typos] of Object.entries(TYPO_PATTERNS)) {
    if (typos.includes(lowerDomain)) {
      const result = {
        original: domain,
        suggested: correctDomain,
        confidence: 0.95
        // High confidence for known typo patterns
      };
      await cacheStore.set(cacheKey, result);
      return result;
    }
  }
  let bestMatch = null;
  const threshold = getSimilarityThreshold(lowerDomain);
  for (const commonDomain of domainsToCheck) {
    const similarity = stringSimilarityJs.stringSimilarity(lowerDomain, commonDomain.toLowerCase());
    if (similarity >= threshold) {
      if (!bestMatch || similarity > bestMatch.similarity) {
        bestMatch = { domain: commonDomain, similarity };
      }
    }
  }
  if (!bestMatch) {
    for (const commonDomain of domainsToCheck) {
      if (Math.abs(lowerDomain.length - commonDomain.length) <= 2) {
        const similarity = stringSimilarityJs.stringSimilarity(lowerDomain, commonDomain.toLowerCase());
        if (similarity >= 0.7) {
          if (!bestMatch || similarity > bestMatch.similarity) {
            bestMatch = { domain: commonDomain, similarity };
          }
        }
      }
    }
  }
  if (bestMatch) {
    if (bestMatch.domain.charAt(0) !== lowerDomain.charAt(0) && bestMatch.similarity < 0.9) {
      await cacheStore.set(cacheKey, null);
      return null;
    }
    const result = {
      original: domain,
      suggested: bestMatch.domain,
      confidence: bestMatch.similarity
    };
    await cacheStore.set(cacheKey, result);
    return result;
  }
  await cacheStore.set(cacheKey, null);
  return null;
}
function suggestDomain(params) {
  const { domain, customMethod, commonDomains } = params;
  if (!domain || domain.length < 3) {
    return null;
  }
  if (customMethod) {
    try {
      return customMethod(domain);
    } catch (error) {
      console.warn("Custom domain suggestion method failed, falling back to default:", error);
    }
  }
  return defaultDomainSuggestionMethod(domain, commonDomains);
}
async function suggestEmailDomain(email, commonDomains, cache) {
  if (!email || !email.includes("@")) {
    return null;
  }
  const [localPart, domain] = email.split("@");
  if (!domain || !localPart) {
    return null;
  }
  const suggestion = await defaultDomainSuggestionMethodAsync(domain, commonDomains, cache);
  if (suggestion) {
    return {
      original: email,
      suggested: `${localPart}@${suggestion.suggested}`,
      confidence: suggestion.confidence
    };
  }
  return null;
}
function isCommonDomain(domain, commonDomains) {
  const domainsToCheck = commonDomains || commonEmailDomains;
  return domainsToCheck.includes(domain.toLowerCase());
}
function getDomainSimilarity(domain1, domain2) {
  return stringSimilarityJs.stringSimilarity(domain1.toLowerCase(), domain2.toLowerCase());
}

async function isValidEmailDomain(emailOrDomain, cache) {
  let [localPart, emailDomain] = (emailOrDomain === null || emailOrDomain === void 0 ? void 0 : emailOrDomain.split("@")) || [];
  if (!emailDomain) {
    emailDomain = localPart;
  }
  if (!emailDomain) {
    return false;
  }
  const cacheStore = getCacheStore(cache, "domainValid");
  const cached = await cacheStore.get(emailDomain);
  if (cached !== null && cached !== void 0) {
    return cached.isValid;
  }
  try {
    const isValidResult = psl.isValid(emailDomain) || false;
    const richResult = {
      isValid: isValidResult,
      hasMX: false,
      // MX not checked in this function
      checkedAt: Date.now()
    };
    await cacheStore.set(emailDomain, richResult);
    return isValidResult;
  } catch (validationError) {
    const errorResult = {
      isValid: false,
      hasMX: false,
      checkedAt: Date.now()
    };
    await cacheStore.set(emailDomain, errorResult);
    return false;
  }
}
function isValidEmail(emailAddress) {
  if (!emailAddress || typeof emailAddress !== "string") {
    return false;
  }
  const re = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
  const emailLower = emailAddress.toLowerCase();
  if (emailLower.indexOf(".+") !== -1)
    return false;
  if (emailLower.indexOf("..") !== -1)
    return false;
  if (emailLower.startsWith(".") || emailLower.endsWith("."))
    return false;
  const parts = emailAddress.split("@");
  if (parts.length !== 2)
    return false;
  const [localPart, domain] = parts;
  if (!localPart || !domain)
    return false;
  if (localPart.length > 64)
    return false;
  if (domain.length > 253)
    return false;
  return re.test(emailLower);
}

async function resolveMxRecords(params) {
  const { domain, cache, logger } = params;
  const log = logger || (() => {
  });
  const cacheStore = getCacheStore(cache, "mx");
  const cached = await cacheStore.get(domain);
  if (cached !== null && cached !== void 0) {
    log(`[resolveMxRecords] Cache hit for ${domain}: ${cached === null || cached === void 0 ? void 0 : cached.length} MX records`);
    return cached;
  }
  log(`[resolveMxRecords] Performing DNS MX lookup for ${domain}`);
  try {
    const records = await node_dns.promises.resolveMx(domain);
    records === null || records === void 0 ? void 0 : records.sort((a, b) => {
      if (a.priority < b.priority) {
        return -1;
      }
      if (a.priority > b.priority) {
        return 1;
      }
      return 0;
    });
    const exchanges = records === null || records === void 0 ? void 0 : records.map((record) => record.exchange);
    log(`[resolveMxRecords] Found ${exchanges === null || exchanges === void 0 ? void 0 : exchanges.length} MX records for ${domain}: [${exchanges === null || exchanges === void 0 ? void 0 : exchanges.join(", ")}]`);
    await cacheStore.set(domain, exchanges);
    log(`[resolveMxRecords] Cached ${exchanges === null || exchanges === void 0 ? void 0 : exchanges.length} MX records for ${domain}`);
    return exchanges;
  } catch (error) {
    log(`[resolveMxRecords] MX lookup failed for ${domain}, caching empty result`);
    await cacheStore.set(domain, []);
    throw error;
  }
}

const nameSeparator = [".", "_", "-"];
const commonNameSuffixes = [
  "mail",
  "email",
  "contact",
  "info",
  "admin",
  "support",
  "sales",
  "help",
  "noreply",
  "no-reply",
  "donotreply",
  "notifications",
  "alerts"
];
const contextualSuffixes = [
  "dev",
  "company",
  "team",
  "group",
  "office",
  "work",
  "personal",
  "home",
  "private",
  "business",
  "corp",
  "inc",
  "ltd",
  "llc",
  "org"
];
const COMMON_TITLES = [
  "mr",
  "mrs",
  "ms",
  "miss",
  "dr",
  "prof",
  "sir",
  "lord",
  "lady",
  "captain",
  "rev",
  "father",
  "sister"
];
const commonFirstName = /* @__PURE__ */ new Set([
  // English names
  "james",
  "john",
  "robert",
  "michael",
  "william",
  "david",
  "richard",
  "joseph",
  "thomas",
  "charles",
  "christopher",
  "daniel",
  "matthew",
  "kenneth",
  "anthony",
  "mark",
  "donald",
  "steven",
  "brian",
  "paul",
  "andrew",
  "joshua",
  "kevin",
  "eric",
  "stephen",
  "benjamin",
  "samuel",
  "gregory",
  "frank",
  "alexander",
  "mary",
  "patricia",
  "jennifer",
  "linda",
  "elizabeth",
  "barbara",
  "susan",
  "jessica",
  "sarah",
  "margaret",
  "karen",
  "lisa",
  "nancy",
  "betty",
  "helen",
  "sandra",
  "donna",
  "carol",
  "ruth",
  "sharon",
  "michelle",
  "laura",
  "kimberly",
  "deborah",
  "amy",
  "angela",
  "ashley",
  "brenda",
  "catherine",
  "emma",
  "olivia",
  "sophia",
  "ava",
  "isabella",
  "mia",
  "charlotte",
  "amelia",
  "harper",
  "evelyn",
  "abigail",
  // Spanish/Latin names
  "juan",
  "carlos",
  "jose",
  "luis",
  "antonio",
  "francisco",
  "manuel",
  "miguel",
  "pedro",
  "jorge",
  "maria",
  "carmen",
  "isabel",
  "ana",
  "lucia",
  "sofia",
  "valentina",
  "camila",
  "gabriela",
  "daniela",
  // Arabic names
  "mohammed",
  "mohamed",
  "ahmed",
  "ali",
  "hassan",
  "ibrahim",
  "omar",
  "khalid",
  "abdullah",
  "yusuf",
  "fatima",
  "aisha",
  "zainab",
  "maryam",
  "khadija",
  "sara",
  "layla",
  "noor",
  "hana",
  "amira",
  // Asian names
  "wei",
  "jing",
  "chen",
  "li",
  "wang",
  "zhang",
  "liu",
  "yang",
  "huang",
  "zhao",
  "satoshi",
  "takeshi",
  "hiroshi",
  "yuki",
  "akira",
  "kenji",
  "taro",
  "naoko",
  "yuko",
  "keiko",
  "raj",
  "arjun",
  "rohit",
  "amit",
  "vikram",
  "priya",
  "anjali",
  "neha",
  "pooja",
  "divya",
  // European names
  "jean",
  "pierre",
  "jacques",
  "marie",
  "francois",
  "giuseppe",
  "marco",
  "alessandro",
  "francesca",
  "giulia",
  "hans",
  "klaus",
  "peter",
  "anna",
  "katrin",
  "vladimir",
  "sergey",
  "olga",
  "natasha",
  "ivan",
  // African names
  "kwame",
  "kofi",
  "amara",
  "fatou",
  "aminata",
  "ousmane",
  "amadou",
  "ibrahim",
  "sekou",
  "mariama"
]);
const commonLastName = /* @__PURE__ */ new Set([
  // English surnames
  "smith",
  "johnson",
  "williams",
  "brown",
  "jones",
  "garcia",
  "miller",
  "davis",
  "rodriguez",
  "martinez",
  "hernandez",
  "lopez",
  "gonzalez",
  "wilson",
  "anderson",
  "thomas",
  "taylor",
  "moore",
  "jackson",
  "martin",
  "lee",
  "thompson",
  "white",
  "harris",
  "clark",
  "lewis",
  "robinson",
  "walker",
  "young",
  "king",
  // Spanish surnames
  "sanchez",
  "ramirez",
  "torres",
  "flores",
  "rivera",
  "gomez",
  "diaz",
  "reyes",
  "morales",
  "cruz",
  "ortiz",
  "gutierrez",
  "chavez",
  "ruiz",
  "alvarez",
  "castillo",
  "jimenez",
  "vasquez",
  "romero",
  "herrera",
  // Asian surnames
  "wang",
  "li",
  "zhang",
  "chen",
  "liu",
  "yang",
  "huang",
  "zhao",
  "wu",
  "zhou",
  "tanaka",
  "watanabe",
  "ito",
  "yamamoto",
  "nakamura",
  "kim",
  "park",
  "choi",
  "jung",
  "kang",
  "patel",
  "sharma",
  "kumar",
  "singh",
  "gupta",
  "khan",
  "reddy",
  "rao",
  "mehta",
  "shah",
  // Other common surnames
  "nguyen",
  "tran",
  "pham",
  "murphy",
  "kelly",
  "sullivan",
  "walsh",
  "connor",
  "ryan",
  "byrne",
  "schmidt",
  "mueller",
  "schneider",
  "fischer",
  "weber",
  "rossi",
  "russo",
  "ferrari",
  "bianchi",
  "romano",
  "silva",
  "santos",
  "ferreira",
  "pereira",
  "costa",
  "ivanov",
  "petrov",
  "sidorov",
  "smirnov",
  "volkov"
]);
function isYearLike(str) {
  return /^(19|20)\d{2}$/.test(str);
}
function isKnownFirstName(str) {
  return commonFirstName.has(str.toLowerCase());
}
function isKnownLastName(str) {
  return commonLastName.has(str.toLowerCase());
}
function isTitle(str) {
  return COMMON_TITLES.includes(str.toLowerCase().replace(".", ""));
}
function getFirstNameScore(str) {
  const lower = str.toLowerCase();
  if (commonFirstName.has(lower))
    return 1;
  if (commonLastName.has(lower))
    return 0.3;
  if (str.length >= 2 && str.length <= 15 && /^[a-zA-Z]+$/.test(str))
    return 0.5;
  return 0;
}
function getLastNameScore(str) {
  const lower = str.toLowerCase();
  if (commonLastName.has(lower))
    return 1;
  if (commonFirstName.has(lower))
    return 0.3;
  if (str.length >= 2 && str.length <= 20 && /^[a-zA-Z]+$/.test(str))
    return 0.5;
  return 0;
}
function capitalizeName(str) {
  if (!str)
    return "";
  const specialPatterns = [
    {
      pattern: /^(o')([a-z]+)/i,
      format: (m) => `O'${m[2].charAt(0).toUpperCase()}${m[2].slice(1).toLowerCase()}`
    },
    {
      pattern: /^(mac)([a-z]+)/i,
      format: (m) => `Mac${m[2].charAt(0).toUpperCase()}${m[2].slice(1).toLowerCase()}`
    },
    {
      pattern: /^(mc)([a-z]+)/i,
      format: (m) => `Mc${m[2].charAt(0).toUpperCase()}${m[2].slice(1).toLowerCase()}`
    },
    {
      pattern: /^(van|von|de|del|dela|da|di|le|la|du|den|der|ter|ten)(\s+)([a-z]+)/i,
      format: (m) => m[1].toLowerCase() + m[2] + m[3].charAt(0).toUpperCase() + m[3].slice(1).toLowerCase()
    }
  ];
  for (const { pattern, format } of specialPatterns) {
    const match = str.match(pattern);
    if (match) {
      return format(match);
    }
  }
  if (str.includes("-")) {
    return str.split("-").map((part) => part.charAt(0).toUpperCase() + part.slice(1).toLowerCase()).join("-");
  }
  if (str.includes("'") && str.indexOf("'") > 0) {
    const parts = str.split("'");
    return parts.map((part, i) => i === 0 ? part.charAt(0).toUpperCase() + part.slice(1).toLowerCase() : part.charAt(0).toUpperCase() + part.slice(1).toLowerCase()).join("'");
  }
  if (/^[a-zA-Z]/.test(str)) {
    return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
  }
  return str;
}
function parseCompositeNamePart(part) {
  const hasNumbers = /\d/.test(part);
  let confidence = 1;
  let cleaned = part;
  const leetSpeakMap = {
    "0": "o",
    "1": "i",
    "3": "e",
    "4": "a",
    "5": "s",
    "7": "t",
    "8": "b"
  };
  if (hasNumbers && part.length >= 3) {
    let leetCleaned = part.toLowerCase();
    for (const [num, letter] of Object.entries(leetSpeakMap)) {
      leetCleaned = leetCleaned.replace(new RegExp(num, "g"), letter);
    }
    if (isKnownFirstName(leetCleaned) || isKnownLastName(leetCleaned)) {
      cleaned = leetCleaned;
      confidence = 0.7;
    }
  }
  if (cleaned === part) {
    const withoutYear = part.replace(/(19|20)\d{2}$/, "");
    if (withoutYear !== part && withoutYear.length >= 2) {
      cleaned = withoutYear;
      confidence = 0.8;
    } else {
      const pureAlpha = part.replace(/\d+/g, "");
      if (pureAlpha.length >= 2) {
        cleaned = pureAlpha;
        confidence = hasNumbers ? 0.85 : 1;
      } else if (pureAlpha.length === 1 && /^[a-zA-Z]$/.test(pureAlpha)) {
        cleaned = pureAlpha;
        confidence = 0.6;
      } else {
        const baseMatch2 = part.match(/^([a-zA-Z]+)\d*$/);
        if (baseMatch2) {
          cleaned = baseMatch2[1];
          confidence = 0.75;
        } else {
          cleaned = part;
          confidence = 0.4;
        }
      }
    }
  }
  if (cleaned !== part) {
    if (isKnownFirstName(cleaned) || isKnownLastName(cleaned)) {
      confidence = Math.min(1, confidence + 0.2);
    }
  }
  const baseMatch = part.match(/^([a-zA-Z]+[a-zA-Z0-9]*?)\d*$/);
  const base = baseMatch ? baseMatch[1] : part;
  return { base, hasNumbers, cleaned, confidence };
}
function isLikelyName(str, allowNumbers = false, allowSingleLetter = false) {
  if (!str)
    return false;
  if (str.length < 2 && !allowSingleLetter)
    return false;
  if (str.length === 1 && allowSingleLetter && /^[a-zA-Z]$/.test(str))
    return true;
  if (commonNameSuffixes.includes(str.toLowerCase()))
    return false;
  if (allowNumbers) {
    if (/^\d+$/.test(str))
      return false;
    return true;
  }
  const digitCount = (str.match(/\d/g) || []).length;
  if (digitCount > str.length * 0.3)
    return false;
  return true;
}
function defaultNameDetectionMethod(email) {
  if (!email || !email.includes("@")) {
    return null;
  }
  const [localPart] = email.split("@");
  if (!localPart) {
    return null;
  }
  const localWithoutAlias = localPart.split("+")[0];
  const cleanedLocal = localWithoutAlias;
  if (cleanedLocal.toLowerCase() === "no-reply" || cleanedLocal.toLowerCase() === "noreply") {
    return null;
  }
  let firstName;
  let lastName;
  let confidence = 0;
  const camelCasePatterns = [
    /^([a-z]+)([A-Z][a-z]+)$/,
    // johnDoe
    /^([A-Z][a-z]+)([A-Z][a-z]+)$/,
    // JohnSmith
    /^([a-z]+)([A-Z][a-z]+)([A-Z][a-z]+)$/
    // johnMcDonald
  ];
  for (const pattern of camelCasePatterns) {
    const match = cleanedLocal.match(pattern);
    if (match) {
      if (match.length === 3) {
        const [, first, last] = match;
        if (isLikelyName(first) && isLikelyName(last)) {
          const firstScore = getFirstNameScore(first);
          const lastScore = getLastNameScore(last);
          firstName = capitalizeName(first);
          lastName = capitalizeName(last);
          confidence = 0.7 + (firstScore + lastScore) * 0.15;
          break;
        }
      } else if (match.length === 4) {
        const [, first, middle, last] = match;
        const combined = middle + last;
        if (isLikelyName(first)) {
          firstName = capitalizeName(first);
          lastName = capitalizeName(combined);
          confidence = 0.75;
          break;
        }
      }
    }
  }
  if (!firstName && !lastName) {
    for (const separator of nameSeparator) {
      if (cleanedLocal.includes(separator)) {
        const parts = cleanedLocal.split(separator).filter((p) => p.length > 0);
        if (parts.length === 2) {
          const [first, last] = parts;
          if (isTitle(first)) {
            continue;
          }
          const firstParsed = parseCompositeNamePart(first);
          const lastParsed = parseCompositeNamePart(last);
          const firstNameScore = getFirstNameScore(firstParsed.cleaned);
          const lastNameScore = getLastNameScore(lastParsed.cleaned);
          const reverseScore = getLastNameScore(firstParsed.cleaned) + getFirstNameScore(lastParsed.cleaned);
          const normalScore = firstNameScore + lastNameScore;
          const firstLikely = isLikelyName(first, true, true);
          const lastLikely = isLikelyName(last, true, true);
          const oneIsSingleLetter = first.length === 1 || last.length === 1;
          const otherIsLongEnough = oneIsSingleLetter ? first.length === 1 ? last.length >= 2 : first.length >= 2 : true;
          const bothPartsValid = firstLikely && lastLikely && otherIsLongEnough;
          if (bothPartsValid) {
            const useReversed = reverseScore > normalScore * 1.2;
            if (firstParsed.hasNumbers || lastParsed.hasNumbers) {
              const cleanedFirst = firstParsed.cleaned;
              const cleanedLast = lastParsed.cleaned;
              if (isLikelyName(cleanedFirst, false, true) && isLikelyName(cleanedLast, false, true)) {
                if (useReversed) {
                  firstName = capitalizeName(cleanedLast);
                  lastName = capitalizeName(cleanedFirst);
                } else {
                  firstName = capitalizeName(cleanedFirst);
                  lastName = capitalizeName(cleanedLast);
                }
                const baseConfidence = 0.7;
                const separatorBonus = separator === "." ? 0.15 : separator === "_" ? 0.05 : 0;
                const nameRecognitionBonus = normalScore > 0.5 ? 0.1 : 0;
                const cleaningPenalty = (firstParsed.confidence + lastParsed.confidence) / 2 - 1;
                confidence = Math.min(0.95, baseConfidence + separatorBonus + nameRecognitionBonus + cleaningPenalty);
              } else {
                firstName = capitalizeName(first);
                lastName = capitalizeName(last);
                confidence = 0.5;
              }
            } else {
              if (useReversed) {
                firstName = capitalizeName(last);
                lastName = capitalizeName(first);
              } else {
                firstName = capitalizeName(first);
                lastName = capitalizeName(last);
              }
              const baseConfidence = 0.75;
              const separatorBonus = separator === "." ? 0.15 : separator === "_" ? 0.05 : 0;
              const nameRecognitionBonus = Math.min(0.15, normalScore * 0.15);
              confidence = Math.min(0.95, baseConfidence + separatorBonus + nameRecognitionBonus);
            }
            break;
          }
        } else if (parts.length === 3) {
          const [first, middle, last] = parts;
          const firstParsed = parseCompositeNamePart(first);
          const middleParsed = parseCompositeNamePart(middle);
          const lastParsed = parseCompositeNamePart(last);
          const isLastSuffix = commonNameSuffixes.includes(last.toLowerCase()) || contextualSuffixes.includes(last.toLowerCase()) || isYearLike(last);
          if (isLastSuffix) {
            if (isLikelyName(first, true, true) && isLikelyName(middle, true, true)) {
              const cleanedFirst = firstParsed.hasNumbers ? firstParsed.cleaned : first;
              const cleanedMiddle = middleParsed.hasNumbers ? middleParsed.cleaned : middle;
              if (isLikelyName(cleanedFirst, false, true) && isLikelyName(cleanedMiddle, false, true)) {
                firstName = capitalizeName(cleanedFirst);
                lastName = capitalizeName(cleanedMiddle);
                confidence = 0.7;
              } else {
                firstName = capitalizeName(first);
                lastName = capitalizeName(middle);
                confidence = 0.65;
              }
              break;
            }
          } else if (isLikelyName(first, true, true) && isLikelyName(last, true, true)) {
            const cleanedFirst = firstParsed.hasNumbers ? firstParsed.cleaned : first;
            const cleanedLast = lastParsed.hasNumbers ? lastParsed.cleaned : last;
            if (isLikelyName(cleanedFirst, false, true) && isLikelyName(cleanedLast, false, true)) {
              firstName = capitalizeName(cleanedFirst);
              lastName = capitalizeName(cleanedLast);
              confidence = 0.75;
            } else {
              firstName = capitalizeName(first);
              lastName = capitalizeName(last);
              confidence = 0.55;
            }
            break;
          }
        } else if (parts.length > 3) {
          const firstPart = parts[0];
          const lastPartLower = parts[parts.length - 1].toLowerCase();
          const isLastPartSuffix = commonNameSuffixes.includes(lastPartLower) || contextualSuffixes.includes(lastPartLower) || isYearLike(parts[parts.length - 1]);
          const effectiveLastIndex = isLastPartSuffix ? parts.length - 2 : parts.length - 1;
          const lastToUse = effectiveLastIndex >= 0 ? parts[effectiveLastIndex] : null;
          if (lastToUse && isLikelyName(firstPart, true, true) && isLikelyName(lastToUse, true, true)) {
            const firstParsed = parseCompositeNamePart(firstPart);
            const lastParsed = parseCompositeNamePart(lastToUse);
            const cleanedFirst = firstParsed.hasNumbers ? firstParsed.cleaned : firstPart;
            const cleanedLast = lastParsed.hasNumbers ? lastParsed.cleaned : lastToUse;
            if (isLikelyName(cleanedFirst, false, true) && isLikelyName(cleanedLast, false, true)) {
              firstName = capitalizeName(cleanedFirst);
              lastName = capitalizeName(cleanedLast);
              confidence = 0.6;
            } else {
              firstName = capitalizeName(firstPart);
              lastName = capitalizeName(lastToUse);
              confidence = 0.5;
            }
            break;
          }
        }
      }
    }
  }
  if (!firstName && !lastName) {
    const underscoreMatch = cleanedLocal.match(/^([A-Z_]+)$/);
    if (underscoreMatch) {
      const parts = cleanedLocal.toLowerCase().split("_").filter((p) => p.length > 0);
      if (parts.length === 2) {
        const [first, last] = parts;
        if (isLikelyName(first) && isLikelyName(last)) {
          firstName = capitalizeName(first);
          lastName = capitalizeName(last);
          confidence = 0.65;
        }
      }
    }
  }
  if (!firstName && !lastName) {
    const parsed = parseCompositeNamePart(cleanedLocal);
    if (isLikelyName(cleanedLocal, true, false)) {
      if (/^[a-zA-Z]+$/.test(cleanedLocal)) {
        const nameScore = Math.max(getFirstNameScore(cleanedLocal), getLastNameScore(cleanedLocal));
        if (getFirstNameScore(cleanedLocal) >= getLastNameScore(cleanedLocal)) {
          firstName = capitalizeName(cleanedLocal);
        } else {
          lastName = capitalizeName(cleanedLocal);
        }
        confidence = 0.4 + nameScore * 0.3;
      } else if (parsed.hasNumbers && parsed.cleaned.length >= 2) {
        const cleanedScore = Math.max(getFirstNameScore(parsed.cleaned), getLastNameScore(parsed.cleaned));
        if (getFirstNameScore(parsed.cleaned) >= getLastNameScore(parsed.cleaned)) {
          firstName = capitalizeName(parsed.cleaned);
        } else {
          lastName = capitalizeName(parsed.cleaned);
        }
        confidence = 0.3 + cleanedScore * 0.2 + parsed.confidence * 0.2;
      }
    }
  }
  if (!firstName && !lastName) {
    return null;
  }
  return {
    firstName,
    lastName,
    confidence: Math.max(0, Math.min(1, confidence))
    // Ensure confidence is between 0 and 1
  };
}
function detectNameFromEmail(params) {
  const { email, customMethod } = params;
  if (!email || !email.includes("@")) {
    return null;
  }
  if (customMethod) {
    try {
      return customMethod(email);
    } catch (error) {
      console.warn("Custom name detection method failed, falling back to default:", error);
    }
  }
  return defaultNameDetectionMethod(email);
}
function cleanNameForAlgorithm(name) {
  if (!name)
    return "";
  let cleaned = name.replace(/[._*]/g, "");
  cleaned = cleaned.replace(/\s+/g, " ").trim();
  if (!cleaned) {
    return name;
  }
  return cleaned;
}
function detectNameForAlgorithm(email) {
  const detectedName = detectName(email);
  if (!detectedName) {
    return null;
  }
  const cleanedFirstName = detectedName.firstName ? cleanNameForAlgorithm(detectedName.firstName) : void 0;
  const cleanedLastName = detectedName.lastName ? cleanNameForAlgorithm(detectedName.lastName) : void 0;
  if (!cleanedFirstName && !cleanedLastName) {
    return null;
  }
  return {
    firstName: cleanedFirstName,
    lastName: cleanedLastName,
    confidence: detectedName.confidence * 0.95
    // Slightly reduce confidence due to cleaning
  };
}
function detectName(email) {
  return detectNameFromEmail({ email });
}

exports.VerificationErrorCode = void 0;
(function(VerificationErrorCode2) {
  VerificationErrorCode2["invalidFormat"] = "INVALID_FORMAT";
  VerificationErrorCode2["invalidDomain"] = "INVALID_DOMAIN";
  VerificationErrorCode2["noMxRecords"] = "NO_MX_RECORDS";
  VerificationErrorCode2["smtpConnectionFailed"] = "SMTP_CONNECTION_FAILED";
  VerificationErrorCode2["smtpTimeout"] = "SMTP_TIMEOUT";
  VerificationErrorCode2["mailboxNotFound"] = "MAILBOX_NOT_FOUND";
  VerificationErrorCode2["mailboxFull"] = "MAILBOX_FULL";
  VerificationErrorCode2["networkError"] = "NETWORK_ERROR";
  VerificationErrorCode2["disposableEmail"] = "DISPOSABLE_EMAIL";
  VerificationErrorCode2["freeEmailProvider"] = "FREE_EMAIL_PROVIDER";
})(exports.VerificationErrorCode || (exports.VerificationErrorCode = {}));
exports.EmailProvider = void 0;
(function(EmailProvider2) {
  EmailProvider2["gmail"] = "gmail";
  EmailProvider2["hotmailB2b"] = "hotmail_b2b";
  EmailProvider2["hotmailB2c"] = "hotmail_b2c";
  EmailProvider2["proofpoint"] = "proofpoint";
  EmailProvider2["mimecast"] = "mimecast";
  EmailProvider2["yahoo"] = "yahoo";
  EmailProvider2["everythingElse"] = "everything_else";
})(exports.EmailProvider || (exports.EmailProvider = {}));
function parseSmtpError(errorMessage) {
  const lowerError = errorMessage.toLowerCase();
  const networkErrorPatterns = [
    "etimedout",
    "econnrefused",
    "enotfound",
    "econnreset",
    "socket hang up",
    "connection_timeout",
    "socket_timeout",
    "connection_error",
    "connection_closed"
  ];
  const isNetworkError = networkErrorPatterns.some((pattern) => lowerError.includes(pattern));
  if (isNetworkError) {
    return {
      isDisabled: false,
      hasFullInbox: false,
      isInvalid: true,
      isCatchAll: false
    };
  }
  const disabledPatterns = [
    "account disabled",
    "account is disabled",
    "user disabled",
    "user is disabled",
    "account locked",
    "account is locked",
    "user blocked",
    "user is blocked",
    "mailbox disabled",
    "delivery not authorized",
    "message rejected",
    "access denied",
    "permission denied",
    "recipient unknown",
    "recipient address rejected",
    "user unknown",
    "address unknown",
    "invalid recipient",
    "not a valid recipient",
    "recipient does not exist",
    "no such user",
    "user does not exist",
    "mailbox unavailable",
    "recipient unavailable",
    "address rejected",
    "550",
    "551",
    "553",
    "not_found",
    "ambiguous"
  ];
  const fullInboxPatterns = [
    "mailbox full",
    "inbox full",
    "quota exceeded",
    "over quota",
    "storage limit exceeded",
    "message too large",
    "insufficient storage",
    "mailbox over quota",
    "over the quota",
    "mailbox size limit exceeded",
    "account over quota",
    "storage space",
    "overquota",
    "452",
    "552",
    "over_quota"
  ];
  const catchAllPatterns = [
    "accept all mail",
    "catch-all",
    "catchall",
    "wildcard",
    "accepts any recipient",
    "recipient address accepted"
  ];
  const rateLimitPatterns = [
    "receiving mail at a rate that",
    "rate limit",
    "too many messages",
    "temporarily rejected",
    "try again later",
    "greylisted",
    "greylist",
    "deferring",
    "temporarily deferred",
    "421",
    "450",
    "451",
    "temporary_failure"
  ];
  const isDisabled = disabledPatterns.some((pattern) => lowerError.includes(pattern)) || lowerError.startsWith("550") || lowerError.startsWith("551") || lowerError.startsWith("553");
  const hasFullInbox = fullInboxPatterns.some((pattern) => lowerError.includes(pattern)) || lowerError.startsWith("452") || lowerError.startsWith("552");
  const isCatchAll = catchAllPatterns.some((pattern) => lowerError.includes(pattern));
  const isInvalid = !isDisabled && !hasFullInbox && !isCatchAll && !rateLimitPatterns.some((pattern) => lowerError.includes(pattern)) && !lowerError.startsWith("421") && !lowerError.startsWith("450") && !lowerError.startsWith("451");
  return {
    isDisabled,
    hasFullInbox,
    isInvalid,
    isCatchAll
  };
}
exports.SMTPStep = void 0;
(function(SMTPStep2) {
  SMTPStep2["greeting"] = "GREETING";
  SMTPStep2["ehlo"] = "EHLO";
  SMTPStep2["helo"] = "HELO";
  SMTPStep2["startTls"] = "STARTTLS";
  SMTPStep2["mailFrom"] = "MAIL_FROM";
  SMTPStep2["rcptTo"] = "RCPT_TO";
  SMTPStep2["vrfy"] = "VRFY";
  SMTPStep2["quit"] = "QUIT";
})(exports.SMTPStep || (exports.SMTPStep = {}));

function isIPAddress(host) {
  const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
  if (ipv4Regex.test(host)) {
    const octets = host.split(".");
    return octets.every((octet) => parseInt(octet, 10) <= 255);
  }
  const ipv6Regex = /^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,7}:$|^(?:[0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,5}(?::[0-9a-fA-F]{1,4}){1,2}$|^(?:[0-9a-fA-F]{1,4}:){1,4}(?::[0-9a-fA-F]{1,4}){1,3}$|^(?:[0-9a-fA-F]{1,4}:){1,3}(?::[0-9a-fA-F]{1,4}){1,4}$|^(?:[0-9a-fA-F]{1,4}:){1,2}(?::[0-9a-fA-F]{1,4}){1,5}$|^[0-9a-fA-F]{1,4}:(?:(?::[0-9a-fA-F]{1,4}){1,6})$|^::1(?::(?::[0-9a-fA-F]{1,4}){1,7})|$|:(?:(?::[0-9a-fA-F]{1,4}){1,7}:)$/;
  return ipv6Regex.test(host);
}
function isHighVolume(smtpReply) {
  return Boolean(smtpReply && /high number of/gi.test(smtpReply));
}
function isOverQuota(smtpReply) {
  return Boolean(smtpReply && /(over quota)/gi.test(smtpReply));
}
function isInvalidMailboxError(smtpReply) {
  return Boolean(smtpReply && /^(510|511|513|550|551|553)/.test(smtpReply) && !/(junk|spam|openspf|spoofing|host|rbl.+blocked)/gi.test(smtpReply));
}
function isMultilineGreet(smtpReply) {
  return Boolean(smtpReply && /^(250|220)-/.test(smtpReply));
}
const DEFAULT_PORTS = [25, 587, 465];
const DEFAULT_TIMEOUT = 3e3;
const DEFAULT_MAX_RETRIES = 1;
const PORT_CONFIGS = {
  25: { tls: false, starttls: true },
  587: { tls: false, starttls: true },
  465: { tls: true, starttls: false }
};
async function verifyMailboxSMTP(params) {
  const { local, domain, mxRecords = [], options = {} } = params;
  const { ports = DEFAULT_PORTS, timeout = DEFAULT_TIMEOUT, maxRetries = DEFAULT_MAX_RETRIES, tls: tlsConfig = true, hostname = "localhost", useVRFY = true, cache, debug = false, sequence } = options;
  const log = debug ? (...args) => console.log("[SMTP]", ...args) : () => {
  };
  const createSmtpResult = (result, port, tlsUsed, mxHost2) => {
    const reason = result === true ? "valid" : result === null ? "ambiguous" : "not_found";
    const parsedError = parseSmtpError(reason);
    return {
      canConnectSmtp: result !== null,
      hasFullInbox: parsedError.hasFullInbox,
      isCatchAll: parsedError.isCatchAll,
      isDeliverable: result === true,
      isDisabled: result === false && parsedError.isDisabled,
      error: result === null ? reason : result === false ? reason : void 0,
      providerUsed: exports.EmailProvider.everythingElse,
      checkedAt: Date.now()
    };
  };
  const createFailureResult = (error) => ({
    canConnectSmtp: false,
    hasFullInbox: false,
    isCatchAll: false,
    isDeliverable: false,
    isDisabled: false,
    error,
    providerUsed: exports.EmailProvider.everythingElse,
    checkedAt: Date.now()
  });
  if (!mxRecords || mxRecords.length === 0) {
    log("No MX records found");
    return {
      smtpResult: createFailureResult("No MX records found"),
      cached: false,
      port: 0,
      portCached: false
    };
  }
  const mxHost = mxRecords[0];
  log(`Verifying ${local}@${domain} via ${mxHost}`);
  const smtpCacheStore = cache ? getCacheStore(cache, "smtp") : null;
  if (smtpCacheStore) {
    let cachedResult;
    try {
      cachedResult = await smtpCacheStore.get(`${mxHost}:${local}@${domain}`);
      if (cachedResult !== void 0 && cachedResult !== null) {
        log(`Using cached SMTP result: ${cachedResult.isDeliverable}`);
        return {
          smtpResult: cachedResult,
          cached: true,
          port: 0,
          portCached: false
        };
      }
    } catch (ignoredError) {
      cachedResult = void 0;
    }
  }
  const smtpPortCacheStore = cache ? getCacheStore(cache, "smtpPort") : null;
  if (smtpPortCacheStore) {
    let cachedPort;
    try {
      cachedPort = await smtpPortCacheStore.get(mxHost);
    } catch (ignoredError) {
      cachedPort = null;
    }
    if (cachedPort) {
      log(`Using cached port: ${cachedPort}`);
      const result = await testSMTPConnection({
        mxHost,
        port: cachedPort,
        local,
        domain,
        timeout,
        tlsConfig,
        hostname,
        useVRFY,
        sequence,
        log
      });
      const smtpResult = createSmtpResult(result);
      if (smtpCacheStore) {
        try {
          await smtpCacheStore.set(`${mxHost}:${local}@${domain}`, smtpResult);
          log(`Cached SMTP result ${result} for ${local}@${domain} via ${mxHost}`);
        } catch (ignoredError) {
        }
      }
      return { smtpResult, cached: false, port: cachedPort, portCached: true };
    }
  }
  for (const port of ports) {
    log(`Testing port ${port}`);
    for (let attempt = 0; attempt < maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = Math.min(200 * 2 ** (attempt - 1), 800);
        log(`Retry ${attempt + 1}, waiting ${delay}ms`);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
      const result = await testSMTPConnection({
        mxHost,
        port,
        local,
        domain,
        timeout,
        tlsConfig,
        hostname,
        useVRFY,
        sequence,
        log
      });
      const smtpResult = createSmtpResult(result);
      if (smtpCacheStore) {
        try {
          await smtpCacheStore.set(`${mxHost}:${local}@${domain}`, smtpResult);
          log(`Cached SMTP result ${result} for ${local}@${domain} via ${mxHost}`);
        } catch (ignoredError) {
        }
      }
      if (result !== null) {
        if (smtpPortCacheStore) {
          try {
            await smtpPortCacheStore.set(mxHost, port);
            log(`Cached port ${port} for ${mxHost}`);
          } catch (ignoredError) {
          }
        }
        return { smtpResult, cached: false, port, portCached: false };
      }
    }
  }
  log("All ports failed");
  return {
    smtpResult: createFailureResult("All SMTP connection attempts failed"),
    cached: false,
    port: 0,
    portCached: false
  };
}
async function testSMTPConnection(params) {
  const { mxHost, port, local, domain, timeout, tlsConfig, hostname, useVRFY, sequence, log } = params;
  const portConfig = PORT_CONFIGS[port] || { tls: false, starttls: false };
  const useTLS = tlsConfig !== false && (portConfig.tls || portConfig.starttls);
  const implicitTLS = portConfig.tls;
  const defaultSequence = {
    steps: [exports.SMTPStep.greeting, exports.SMTPStep.ehlo, exports.SMTPStep.mailFrom, exports.SMTPStep.rcptTo]
  };
  const activeSequence = sequence || defaultSequence;
  if (port === 25) {
    activeSequence.steps = activeSequence.steps.map((step) => step === exports.SMTPStep.ehlo ? exports.SMTPStep.helo : step);
  }
  const tlsOptions = {
    host: mxHost,
    servername: isIPAddress(mxHost) ? void 0 : mxHost,
    // Don't set servername for IP addresses
    rejectUnauthorized: false,
    minVersion: "TLSv1.2",
    ...typeof tlsConfig === "object" ? tlsConfig : {}
  };
  return new Promise((resolve) => {
    let socket;
    let buffer = "";
    let isTLS = implicitTLS;
    let currentStepIndex = 0;
    let resolved = false;
    let supportsSTARTTLS = false;
    let supportsVRFY = false;
    let cleanup = () => {
      if (resolved)
        return;
      resolved = true;
      try {
        socket === null || socket === void 0 ? void 0 : socket.write("QUIT\r\n");
      } catch {
      }
      setTimeout(() => socket === null || socket === void 0 ? void 0 : socket.destroy(), 100);
    };
    const finish = (result, reason) => {
      if (resolved)
        return;
      log(`${port}: ${reason || (result ? "valid" : "invalid")}`);
      cleanup();
      resolve(result);
    };
    const sendCommand = (cmd) => {
      if (resolved)
        return;
      log(`\u2192 ${cmd}`);
      socket === null || socket === void 0 ? void 0 : socket.write(`${cmd}\r
`);
    };
    const nextStep = () => {
      currentStepIndex++;
      if (currentStepIndex >= activeSequence.steps.length) {
        finish(true, "sequence_complete");
        return;
      }
      executeStep(activeSequence.steps[currentStepIndex]);
    };
    const executeStep = (step) => {
      if (resolved)
        return;
      switch (step) {
        case exports.SMTPStep.ehlo:
          sendCommand(`EHLO ${hostname}`);
          break;
        case exports.SMTPStep.helo:
          sendCommand(`HELO ${domain}`);
          break;
        case exports.SMTPStep.greeting:
          break;
        case exports.SMTPStep.startTls:
          sendCommand("STARTTLS");
          break;
        case exports.SMTPStep.mailFrom: {
          const from = activeSequence.from || `<${local}@${domain}>`;
          sendCommand(`MAIL FROM:${from}`);
          break;
        }
        case exports.SMTPStep.rcptTo:
          sendCommand(`RCPT TO:<${local}@${domain}>`);
          break;
        case exports.SMTPStep.vrfy: {
          const vrfyTarget = activeSequence.vrfyTarget || local;
          sendCommand(`VRFY ${vrfyTarget}`);
          break;
        }
        case exports.SMTPStep.quit:
          sendCommand("QUIT");
          break;
      }
    };
    const processResponse = (response) => {
      if (resolved)
        return;
      const code = response.substring(0, 3);
      const isMultiline = response.length > 3 && response[3] === "-";
      log(`\u2190 ${response}`);
      if (isMultilineGreet(response)) {
        return;
      }
      if (isHighVolume(response)) {
        finish(true, "high_volume");
        return;
      }
      if (isOverQuota(response)) {
        finish(false, "over_quota");
        return;
      }
      if (isInvalidMailboxError(response)) {
        finish(false, "not_found");
        return;
      }
      if (isMultiline) {
        const currentStep2 = activeSequence.steps[currentStepIndex];
        if (currentStep2 === exports.SMTPStep.ehlo && code === "250") {
          const upper = response.toUpperCase();
          if (upper.includes("STARTTLS"))
            supportsSTARTTLS = true;
          if (upper.includes("VRFY"))
            supportsVRFY = true;
        }
        if (currentStep2 === exports.SMTPStep.helo && code === "250") {
          const upper = response.toUpperCase();
          if (upper.includes("VRFY"))
            supportsVRFY = true;
        }
        return;
      }
      if (!response.includes("220") && !response.includes("250") && !response.includes("550") && !response.includes("552")) {
        finish(null, "unrecognized_response");
        return;
      }
      const currentStep = activeSequence.steps[currentStepIndex];
      switch (currentStep) {
        case exports.SMTPStep.greeting:
          if (code.startsWith("220")) {
            nextStep();
          } else {
            finish(null, "no_greeting");
          }
          break;
        case exports.SMTPStep.ehlo:
          if (code.startsWith("250")) {
            const hasSTARTTLS = activeSequence.steps.includes(exports.SMTPStep.startTls);
            if (!isTLS && useTLS && supportsSTARTTLS && !implicitTLS && hasSTARTTLS) {
              currentStepIndex = activeSequence.steps.indexOf(exports.SMTPStep.startTls);
              executeStep(exports.SMTPStep.startTls);
            } else {
              nextStep();
            }
          } else {
            finish(null, "ehlo_failed");
          }
          break;
        case exports.SMTPStep.helo:
          if (code.startsWith("250")) {
            nextStep();
          } else {
            finish(null, "helo_failed");
          }
          break;
        case exports.SMTPStep.startTls:
          if (code.startsWith("220")) {
            const plainSocket = socket;
            socket = tls__namespace.connect({
              ...tlsOptions,
              socket: plainSocket,
              servername: isIPAddress(mxHost) ? void 0 : mxHost
            }, () => {
              isTLS = true;
              log("TLS upgraded");
              buffer = "";
              const starttlsIndex = activeSequence.steps.indexOf(exports.SMTPStep.startTls);
              currentStepIndex = starttlsIndex;
              nextStep();
            });
            socket.on("data", handleData);
            socket.on("error", () => finish(null, "tls_error"));
          } else {
            nextStep();
          }
          break;
        case exports.SMTPStep.mailFrom:
          if (code.startsWith("250")) {
            nextStep();
          } else {
            finish(null, "mail_from_rejected");
          }
          break;
        case exports.SMTPStep.rcptTo:
          if (code.startsWith("250") || code.startsWith("251")) {
            finish(true, "valid");
          } else if (code.startsWith("552") || code.startsWith("452")) {
            finish(false, "over_quota");
          } else if (code.startsWith("4")) {
            finish(null, "temporary_failure");
          } else if (useVRFY && supportsVRFY && code.startsWith("5") && activeSequence.steps.includes(exports.SMTPStep.vrfy)) {
            currentStepIndex = activeSequence.steps.indexOf(exports.SMTPStep.vrfy);
            executeStep(exports.SMTPStep.vrfy);
          } else {
            finish(null, "ambiguous");
          }
          break;
        case exports.SMTPStep.vrfy:
          if (code.startsWith("250") || code.startsWith("252")) {
            finish(true, "vrfy_valid");
          } else if (code.startsWith("550")) {
            finish(false, "vrfy_invalid");
          } else {
            finish(null, "vrfy_failed");
          }
          break;
        case exports.SMTPStep.quit:
          if (code.startsWith("221")) {
            finish(null, "quit_received");
          }
          break;
      }
    };
    let handleData = (data) => {
      if (resolved)
        return;
      buffer += data.toString();
      let pos;
      while ((pos = buffer.indexOf("\r\n")) !== -1) {
        const line = buffer.substring(0, pos);
        buffer = buffer.substring(pos + 2);
        processResponse(line.trim());
      }
    };
    if (port < 0 || port > 65535 || !Number.isInteger(port)) {
      finish(null, "invalid_port");
      return;
    }
    if (implicitTLS) {
      const connectOptions = {
        ...tlsOptions,
        port,
        servername: isIPAddress(mxHost) ? void 0 : mxHost
      };
      socket = tls__namespace.connect(connectOptions, () => {
        log(`Connected to ${mxHost}:${port} with TLS`);
        socket.on("data", handleData);
      });
    } else {
      socket = net__namespace.connect({ host: mxHost, port }, () => {
        log(`Connected to ${mxHost}:${port}`);
        socket.on("data", handleData);
      });
    }
    if (activeSequence.steps.length === 0) {
      finish(true, "sequence_complete");
      return;
    }
    const firstStep = activeSequence.steps[0];
    let connectionTimeout;
    let stepTimeout;
    const resetActivityTimeout = () => {
      if (stepTimeout) {
        clearTimeout(stepTimeout);
      }
      stepTimeout = setTimeout(() => {
        if (!resolved) {
          log(`Step timeout after ${timeout}ms of inactivity`);
          finish(null, "step_timeout");
        }
      }, timeout);
    };
    connectionTimeout = setTimeout(() => {
      if (!resolved) {
        log(`Connection timeout after ${timeout}ms`);
        finish(null, "connection_timeout");
      }
    }, timeout);
    if (firstStep !== exports.SMTPStep.greeting) {
      executeStep(firstStep);
    }
    socket.setTimeout(timeout, () => {
      if (!resolved) {
        log(`Socket timeout after ${timeout}ms`);
        finish(null, "socket_timeout");
      }
    });
    socket.on("error", (error) => {
      log(`Socket error: ${error.message}`);
      if (!resolved)
        finish(null, "connection_error");
    });
    socket.on("close", () => {
      if (!resolved) {
        log("Socket closed unexpectedly");
        finish(null, "connection_closed");
      }
    });
    const originalHandleData = handleData;
    handleData = (data) => {
      resetActivityTimeout();
      originalHandleData(data);
    };
    const enhancedCleanup = () => {
      if (resolved)
        return;
      resolved = true;
      if (connectionTimeout)
        clearTimeout(connectionTimeout);
      if (stepTimeout)
        clearTimeout(stepTimeout);
      socket.setTimeout(0);
      try {
        socket === null || socket === void 0 ? void 0 : socket.write("QUIT\r\n");
      } catch {
      }
      setTimeout(() => socket === null || socket === void 0 ? void 0 : socket.destroy(), 100);
    };
    cleanup = enhancedCleanup;
  });
}

const defaultRegex = {
  domainName: "Domain Name: *([^\\s]+)",
  registrar: "Registrar: *(.+)",
  updatedDate: "Updated Date: *(.+)",
  creationDate: "Creat(ed|ion) Date: *(.+)",
  expirationDate: "Expir\\w+ Date: *(.+)",
  status: "Status:\\s*(.+)\\s*\\n",
  dateFormat: "yyyy-MM-dd'T'HH:mm:ss'Z'",
  notFound: "(No match for |Domain not found|NOT FOUND\\s)"
};
const comRegex = { ...defaultRegex, notFound: "No match for " };
const orgRegex = { ...defaultRegex, notFound: "^(NOT FOUND|Domain not found)" };
const auRegex = {
  ...defaultRegex,
  updatedDate: "Last Modified: *(.+)",
  registrar: "Registrar Name: *(.+)",
  rateLimited: "WHOIS LIMIT EXCEEDED",
  notFound: "^NOT FOUND"
};
const usRegex = {
  ...defaultRegex,
  status: "Domain Status: *(.+)",
  expirationDate: "Registrar Registration Expiration Date: *(.+)",
  notFound: "^No Data Found"
};
const ruRegex = {
  ...defaultRegex,
  domainName: "domain: *([^\\s]+)",
  registrar: "registrar: *(.+)",
  expirationDate: "paid-till: *(.+)",
  status: "state: *(.+)",
  notFound: "No entries found"
};
const ukRegex = {
  ...defaultRegex,
  domainName: "Domain name:\\s*([^\\s]+)",
  dateFormat: "dd-MMM-yyyy"
};
const frRegex = {
  ...defaultRegex,
  domainName: "domain: *([^\\s]+)",
  expirationDate: "Expir\\w+ Date:\\s?(.+)",
  updatedDate: "last-update: *(.+)",
  notFound: "(No entries found in |%% NOT FOUND)"
};
const nlRegex = {
  ...defaultRegex,
  notFound: "\\.nl is free",
  rateLimited: "maximum number of requests per second exceeded"
};
const fiRegex = {
  ...defaultRegex,
  domainName: "domain\\.*: *([\\S]+)",
  registrar: "registrar\\.*: *(.*)",
  status: "status\\.*: *([\\S]+)",
  dateFormat: "dd.MM.yyyy HH:mm:ss"
};
const jpRegex = {
  ...defaultRegex,
  domainName: "\\[Domain Name\\]\\s*([^\\s]+)",
  dateFormat: "yyyy/MM/dd",
  notFound: "No match!!"
};
const plRegex = {
  ...defaultRegex,
  domainName: "DOMAIN NAME: *([^\\s]+)\\s+",
  status: "Registration status:\\n\\s*(.+)",
  expirationDate: "renewal date: *(.+)",
  dateFormat: "yyyy.MM.dd HH:mm:ss"
};
const brRegex = {
  ...defaultRegex,
  domainName: "domain: *([^\\s]+)\\n",
  dateFormat: "yyyyMMdd"
};
const euRegex = {
  ...defaultRegex,
  domainName: "Domain: *([^\\n\\r]+)",
  registrar: "Registrar: *\\n *Name: *([^\\n\\r]+)",
  notFound: "Status: AVAILABLE"
};
const eeRegex = {
  ...defaultRegex,
  domainName: "Domain: *[\\n\\r]+\\s*name: *([^\\n\\r]+)",
  status: "Domain: *[\\n\\r]+\\s*name: *[^\\n\\r]+\\s*status: *([^\\n\\r]+)"
};
const krRegex = {
  ...defaultRegex,
  domainName: "Domain Name\\s*: *([^\\s]+)",
  dateFormat: "yyyy. MM. dd.",
  notFound: "The requested domain was not found "
};
const bgRegex = {
  ...defaultRegex,
  domainName: "DOMAIN NAME: *([^\\s]+)",
  status: "registration status:\\s*(.+)",
  notFound: "registration status: available",
  rateLimited: "Query limit exceeded"
};
const deRegex = {
  ...defaultRegex,
  domainName: "Domain: *([^\\s]+)",
  notFound: "Status: *free"
};
const atRegex = {
  ...defaultRegex,
  domainName: "domain: *([^\\s]+)",
  notFound: " nothing found",
  dateFormat: "yyyyMMdd HH:mm:ss",
  rateLimited: "Quota exceeded"
};
const caRegex = {
  ...defaultRegex,
  domainName: "Domain Name: *([^\\s]+)",
  notFound: "Not found: "
};
const beRegex = {
  ...defaultRegex,
  domainName: "Domain:\\s*([^\\s]+)",
  dateFormat: "ddd MMM dd yyyy",
  notFound: "Status:\\s*AVAILABLE"
};
const infoRegex = {
  ...defaultRegex,
  notFound: "^(NOT FOUND|Domain not found)"
};
const kgRegex = {
  ...defaultRegex,
  domainName: "^Domain\\s*([^\\s]+)",
  dateFormat: "ddd MMM dd HH:mm:ss yyyy",
  notFound: "domain is available for registration"
};
const idRegex = {
  ...defaultRegex,
  domainName: "Domain Name:([^\\s]+)",
  dateFormat: "dd-MMM-yyyy HH:mm:ss 'UTC'",
  notFound: "DOMAIN NOT FOUND"
};
const skRegex = {
  ...defaultRegex,
  domainName: "Domain:\\s*([^\\s]+)",
  notFound: "Domain not found"
};
const seRegex = {
  ...defaultRegex,
  domainName: "domain\\.*: *([^\\s]+)",
  notFound: '\\" not found.'
};
const isRegex = {
  ...defaultRegex,
  domainName: "domain\\.*: *([^\\s]+)",
  dateFormat: "MMM dd yyyy",
  notFound: "No entries found for query"
};
const coRegex = {
  ...defaultRegex,
  notFound: "No Data Found"
};
function parseDate(dateStr, format) {
  try {
    dateStr = dateStr.trim();
    if (format === "yyyy-MM-dd'T'HH:mm:ss'Z'" || dateStr.match(/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/)) {
      return new Date(dateStr);
    }
    let date = null;
    switch (format) {
      case "yyyy/MM/dd": {
        const parts = dateStr.split("/");
        if (parts.length === 3) {
          date = /* @__PURE__ */ new Date(`${parts[0]}-${parts[1].padStart(2, "0")}-${parts[2].padStart(2, "0")}`);
        }
        break;
      }
      case "dd-MMM-yyyy": {
        date = new Date(dateStr.replace(/-/g, " "));
        break;
      }
      case "dd.MM.yyyy HH:mm:ss": {
        const [datePart, timePart] = dateStr.split(" ");
        if (datePart) {
          const parts = datePart.split(".");
          if (parts.length === 3) {
            const dateString = `${parts[2]}-${parts[1].padStart(2, "0")}-${parts[0].padStart(2, "0")}`;
            date = /* @__PURE__ */ new Date(`${dateString}${timePart ? `T${timePart}` : ""}`);
          }
        }
        break;
      }
      case "yyyy.MM.dd HH:mm:ss": {
        const [datePart, timePart] = dateStr.split(" ");
        if (datePart) {
          const parts = datePart.split(".");
          if (parts.length === 3) {
            const dateString = `${parts[0]}-${parts[1].padStart(2, "0")}-${parts[2].padStart(2, "0")}`;
            date = /* @__PURE__ */ new Date(`${dateString}${timePart ? `T${timePart}` : ""}`);
          }
        }
        break;
      }
      case "yyyyMMdd": {
        if (dateStr.length === 8) {
          const year = dateStr.substring(0, 4);
          const month = dateStr.substring(4, 6);
          const day = dateStr.substring(6, 8);
          date = /* @__PURE__ */ new Date(`${year}-${month}-${day}`);
        }
        break;
      }
      case "yyyy. MM. dd.": {
        const cleaned = dateStr.replace(/\./g, "").trim();
        const parts = cleaned.split(/\s+/);
        if (parts.length === 3) {
          date = /* @__PURE__ */ new Date(`${parts[0]}-${parts[1].padStart(2, "0")}-${parts[2].padStart(2, "0")}`);
        }
        break;
      }
      case "ddd MMM dd yyyy": {
        const parts = dateStr.split(/\s+/);
        if (parts.length >= 4) {
          date = /* @__PURE__ */ new Date(`${parts[1]} ${parts[2]} ${parts[3]}`);
        }
        break;
      }
      case "ddd MMM dd HH:mm:ss yyyy": {
        const parts = dateStr.split(/\s+/);
        if (parts.length >= 5) {
          date = /* @__PURE__ */ new Date(`${parts[1]} ${parts[2]} ${parts[4]} ${parts[3]}`);
        }
        break;
      }
      case "yyyyMMdd HH:mm:ss": {
        const [datePart, timePart] = dateStr.split(" ");
        if (datePart && datePart.length === 8) {
          const year = datePart.substring(0, 4);
          const month = datePart.substring(4, 6);
          const day = datePart.substring(6, 8);
          date = /* @__PURE__ */ new Date(`${year}-${month}-${day}${timePart ? `T${timePart}` : ""}`);
        }
        break;
      }
      case "dd-MMM-yyyy HH:mm:ss 'UTC'": {
        const cleaned = dateStr.replace(/\s*UTC\s*$/, "");
        date = new Date(cleaned.replace(/-/g, " "));
        break;
      }
      case "MMM dd yyyy": {
        date = new Date(dateStr);
        break;
      }
      default: {
        date = new Date(dateStr);
      }
    }
    if (date && !Number.isNaN(date.getTime())) {
      return date;
    }
    const fallback = new Date(dateStr);
    if (!Number.isNaN(fallback.getTime())) {
      return fallback;
    }
    return null;
  } catch {
    return null;
  }
}
function parseWhoisData({ rawData, domain }) {
  if (!rawData) {
    return { domainName: domain, isAvailable: true };
  }
  const result = { domainName: domain };
  let domainRegex;
  if (domain.endsWith(".com") || domain.endsWith(".net") || domain.endsWith(".name")) {
    domainRegex = comRegex;
  } else if (domain.endsWith(".org") || domain.endsWith(".me") || domain.endsWith(".mobi")) {
    domainRegex = orgRegex;
  } else if (domain.endsWith(".au")) {
    domainRegex = auRegex;
  } else if (domain.endsWith(".ru") || domain.endsWith(".\u0440\u0444") || domain.endsWith(".su")) {
    domainRegex = ruRegex;
  } else if (domain.endsWith(".us") || domain.endsWith(".biz")) {
    domainRegex = usRegex;
  } else if (domain.endsWith(".uk")) {
    domainRegex = ukRegex;
  } else if (domain.endsWith(".fr")) {
    domainRegex = frRegex;
  } else if (domain.endsWith(".nl")) {
    domainRegex = nlRegex;
  } else if (domain.endsWith(".fi")) {
    domainRegex = fiRegex;
  } else if (domain.endsWith(".jp")) {
    domainRegex = jpRegex;
  } else if (domain.endsWith(".pl")) {
    domainRegex = plRegex;
  } else if (domain.endsWith(".br")) {
    domainRegex = brRegex;
  } else if (domain.endsWith(".eu")) {
    domainRegex = euRegex;
  } else if (domain.endsWith(".ee")) {
    domainRegex = eeRegex;
  } else if (domain.endsWith(".kr")) {
    domainRegex = krRegex;
  } else if (domain.endsWith(".bg")) {
    domainRegex = bgRegex;
  } else if (domain.endsWith(".de")) {
    domainRegex = deRegex;
  } else if (domain.endsWith(".at")) {
    domainRegex = atRegex;
  } else if (domain.endsWith(".ca")) {
    domainRegex = caRegex;
  } else if (domain.endsWith(".be")) {
    domainRegex = beRegex;
  } else if (domain.endsWith(".kg")) {
    domainRegex = kgRegex;
  } else if (domain.endsWith(".info")) {
    domainRegex = infoRegex;
  } else if (domain.endsWith(".id")) {
    domainRegex = idRegex;
  } else if (domain.endsWith(".sk")) {
    domainRegex = skRegex;
  } else if (domain.endsWith(".se") || domain.endsWith(".nu")) {
    domainRegex = seRegex;
  } else if (domain.endsWith(".is")) {
    domainRegex = isRegex;
  } else if (domain.endsWith(".co")) {
    domainRegex = coRegex;
  } else {
    domainRegex = defaultRegex;
  }
  Object.keys(domainRegex).forEach((key) => {
    var _a, _b;
    const typedKey = key;
    const pattern = domainRegex[typedKey];
    if (!pattern || typedKey === "dateFormat")
      return;
    let regex = new RegExp(pattern, "i");
    if (typedKey === "status") {
      regex = new RegExp(pattern, "gi");
    }
    const matches = rawData.match(regex);
    if (matches) {
      if (typedKey === "rateLimited") {
        result.rateLimited = true;
        throw new Error("Rate Limited");
      } else if (typedKey === "notFound") {
        result.isAvailable = true;
      } else if (typedKey === "status") {
        result.status = [];
        let match;
        const statusRegex = new RegExp(pattern, "gi");
        match = statusRegex.exec(rawData);
        while (match !== null) {
          if (match[1]) {
            result.status.push(match[1].trim());
          }
          match = statusRegex.exec(rawData);
        }
      } else if (typedKey === "creationDate" || typedKey === "expirationDate" || typedKey === "updatedDate") {
        matches[matches.length - 1];
        let dateStr;
        if (typedKey === "creationDate" && pattern.includes("Creat(ed|ion)")) {
          const creationMatch = rawData.match(/Creat(?:ed|ion) Date:\s*(.+)/i);
          dateStr = (_a = creationMatch === null || creationMatch === void 0 ? void 0 : creationMatch[1]) === null || _a === void 0 ? void 0 : _a.trim();
        } else {
          const extractMatch = rawData.match(new RegExp(pattern, "i"));
          dateStr = (_b = extractMatch === null || extractMatch === void 0 ? void 0 : extractMatch[1]) === null || _b === void 0 ? void 0 : _b.trim();
        }
        if (dateStr) {
          const dateFormat = domainRegex.dateFormat || "yyyy-MM-dd'T'HH:mm:ss'Z'";
          const parsedDate = parseDate(dateStr, dateFormat);
          if (parsedDate) {
            result[typedKey] = parsedDate.toISOString();
          } else {
            result[typedKey] = dateStr;
          }
        }
      } else if (typedKey === "domainName") {
        const match = rawData.match(regex);
        if (match === null || match === void 0 ? void 0 : match[1]) {
          result.domainName = match[1].toLowerCase().trim();
        }
      } else if (typedKey === "registrar") {
        const match = rawData.match(regex);
        if (match === null || match === void 0 ? void 0 : match[1]) {
          result.registrar = match[1].trim();
        }
      }
    }
  });
  if (result.isAvailable === void 0) {
    result.isAvailable = false;
  }
  const patterns = [
    { pattern: /Name Server:\s*(.*)/gi, property: "nameServers", multiple: true },
    { pattern: /Registrar URL:\s*(.*)/i, property: "registrarUrl" },
    { pattern: /Registrar WHOIS Server:\s*(.*)/i, property: "registrarWhoisServer" },
    { pattern: /Registry Domain ID:\s*(.*)/i, property: "registryDomainId" },
    { pattern: /Registrar Abuse Contact Email:\s*(.*)/i, property: "registrarAbuseContactEmail" },
    { pattern: /Registrar Abuse Contact Phone:\s*(.*)/i, property: "registrarAbuseContactPhone" },
    { pattern: /DNSSEC:\s*(.*)/i, property: "dnssec" },
    { pattern: /Registrar IANA ID:\s*(.*)/i, property: "registrarIanaId" }
  ];
  patterns.forEach(({ pattern, property, multiple }) => {
    if (multiple) {
      const matches = Array.from(rawData.matchAll(pattern));
      if (matches.length > 0) {
        result[property] = matches.map((m) => m[1].trim());
      }
    } else {
      const match = rawData.match(pattern);
      if (match === null || match === void 0 ? void 0 : match[1]) {
        result[property] = match[1].trim();
      }
    }
  });
  return result;
}

const whoisServers = {
  com: "whois.verisign-grs.com",
  net: "whois.verisign-grs.com",
  org: "whois.pir.org",
  info: "whois.afilias.net",
  biz: "whois.biz",
  io: "whois.nic.io",
  co: "whois.nic.co",
  uk: "whois.nic.uk",
  de: "whois.denic.de",
  fr: "whois.afnic.fr",
  jp: "whois.jprs.jp",
  au: "whois.auda.org.au",
  ca: "whois.cira.ca",
  eu: "whois.eu",
  nl: "whois.domain-registry.nl",
  ru: "whois.tcinet.ru",
  br: "whois.registro.br",
  cn: "whois.cnnic.cn",
  in: "whois.registry.in",
  me: "whois.nic.me",
  us: "whois.nic.us",
  tv: "whois.nic.tv",
  cc: "whois.nic.cc",
  ws: "whois.website.ws",
  it: "whois.nic.it",
  se: "whois.iis.se",
  no: "whois.norid.no",
  dk: "whois.dk-hostmaster.dk",
  fi: "whois.fi",
  es: "whois.nic.es",
  ch: "whois.nic.ch",
  pl: "whois.dns.pl",
  be: "whois.dns.be",
  at: "whois.nic.at",
  ie: "whois.iedr.ie",
  pt: "whois.dns.pt",
  cz: "whois.nic.cz",
  nz: "whois.srs.net.nz",
  za: "whois.registry.net.za",
  sg: "whois.sgnic.sg",
  hk: "whois.hkirc.hk",
  kr: "whois.kr",
  tw: "whois.twnic.net.tw",
  mx: "whois.mx",
  ar: "whois.nic.ar",
  cl: "whois.nic.cl"
};
function queryWhoisServer(domain, server, timeout = 5e3, debug = false) {
  const log = debug ? console.debug : (..._args) => {
  };
  return new Promise((resolve, reject) => {
    const client = new net__namespace.Socket();
    let data = "";
    log(`[whois] querying ${server} for domain ${domain}`);
    const timer = setTimeout(() => {
      log(`[whois] timeout after ${timeout}ms for ${domain} at ${server}`);
      client.destroy();
      reject(new Error("WHOIS query timeout"));
    }, timeout);
    client.connect(43, server, () => {
      log(`[whois] connected to ${server}, sending query for ${domain}`);
      client.write(`${domain}\r
`);
    });
    client.on("data", (chunk) => {
      const chunkStr = chunk.toString();
      data += chunkStr;
      log(`[whois] received ${chunkStr.length} bytes from ${server}`);
    });
    client.on("close", () => {
      clearTimeout(timer);
      log(`[whois] connection closed, received total ${data.length} bytes from ${server}`);
      resolve(data);
    });
    client.on("error", (err) => {
      clearTimeout(timer);
      log(`[whois] error querying ${server}: ${err.message}`);
      reject(err);
    });
  });
}
async function getWhoisData(domain, timeout = 5e3, debug = false, cache) {
  var _a;
  const log = debug ? console.debug : (..._args) => {
  };
  const cacheKey = `whois:${domain}`;
  const cacheStore = getCacheStore(cache, "whois");
  log(`[whois] getting WHOIS data for ${domain}`);
  const cached = await cacheStore.get(cacheKey);
  if (cached !== null && cached !== void 0) {
    log(`[whois] using cached data for ${domain}`);
    return cached;
  }
  try {
    const tld = (_a = domain.split(".").pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    if (!tld) {
      throw new Error("Invalid domain");
    }
    log(`[whois] extracted TLD: ${tld} for domain: ${domain}`);
    const whoisServer = whoisServers[tld];
    if (!whoisServer) {
      log(`[whois] no specific server for TLD ${tld}, trying IANA`);
      const defaultServer = "whois.iana.org";
      const ianaResponse = await queryWhoisServer(domain, defaultServer, timeout, debug);
      const referMatch = ianaResponse.match(/refer:\s+(\S+)/i);
      if (referMatch === null || referMatch === void 0 ? void 0 : referMatch[1]) {
        const referredServer = referMatch[1];
        log(`[whois] IANA referred to ${referredServer} for ${domain}`);
        const whoisResponse2 = await queryWhoisServer(domain, referredServer, timeout, debug);
        const whoisData3 = parseWhoisData({ rawData: whoisResponse2, domain });
        await cacheStore.set(cacheKey, whoisData3);
        log(`[whois] successfully retrieved and cached WHOIS data from referred server for ${domain}`);
        return whoisData3;
      }
      const whoisData2 = parseWhoisData({ rawData: ianaResponse, domain });
      await cacheStore.set(cacheKey, whoisData2);
      log(`[whois] successfully retrieved and cached WHOIS data from IANA for ${domain}`);
      return whoisData2;
    }
    log(`[whois] using WHOIS server ${whoisServer} for TLD ${tld}`);
    const whoisResponse = await queryWhoisServer(domain, whoisServer, timeout, debug);
    const whoisData = parseWhoisData({ rawData: whoisResponse, domain });
    await cacheStore.set(cacheKey, whoisData);
    log(`[whois] successfully retrieved and cached WHOIS data for ${domain}`);
    return whoisData;
  } catch (ignoredError) {
    log(`[whois] failed to get WHOIS data for ${domain}: ${ignoredError instanceof Error ? ignoredError.message : "Unknown error"}`);
    return null;
  }
}
async function getDomainAge(domain, timeout = 5e3, debug = false, cache) {
  const log = debug ? console.debug : (..._args) => {
  };
  try {
    const cleanDomain = domain.replace(/^https?:\/\//, "").split("/")[0].split("@").pop();
    if (!cleanDomain) {
      log(`[whois] invalid domain format: ${domain}`);
      return null;
    }
    log(`[whois] checking domain age for ${cleanDomain}`);
    if (!psl.isValid(cleanDomain)) {
      log(`[whois] domain validation failed: ${cleanDomain}`);
      return null;
    }
    const whoisData = await getWhoisData(cleanDomain, timeout, debug, cache);
    if (!whoisData || !whoisData.creationDate) {
      log(`[whois] no creation date found for ${cleanDomain}`);
      return null;
    }
    const now = /* @__PURE__ */ new Date();
    const creationDate = new Date(whoisData.creationDate);
    const ageInMilliseconds = now.getTime() - creationDate.getTime();
    const ageInDays = Math.floor(ageInMilliseconds / (1e3 * 60 * 60 * 24));
    const ageInYears = ageInDays / 365.25;
    log(`[whois] calculated age for ${cleanDomain}: ${ageInDays} days (${ageInYears.toFixed(2)} years)`);
    return {
      domain: cleanDomain,
      creationDate,
      ageInDays,
      ageInYears: parseFloat(ageInYears.toFixed(2)),
      expirationDate: whoisData.expirationDate ? new Date(whoisData.expirationDate) : null,
      updatedDate: whoisData.updatedDate ? new Date(whoisData.updatedDate) : null
    };
  } catch (ignoredError) {
    log(`[whois] error getting domain age for ${domain}: ${ignoredError instanceof Error ? ignoredError.message : "Unknown error"}`);
    return null;
  }
}
async function getDomainRegistrationStatus(domain, timeout = 5e3, debug = false, cache) {
  const log = debug ? console.debug : (..._args) => {
  };
  try {
    const cleanDomain = domain.replace(/^https?:\/\//, "").split("/")[0].split("@").pop();
    if (!cleanDomain) {
      log(`[whois] invalid domain format: ${domain}`);
      return null;
    }
    log(`[whois] checking domain registration status for ${cleanDomain}`);
    if (!psl.isValid(cleanDomain)) {
      log(`[whois] domain validation failed: ${cleanDomain}`);
      return null;
    }
    const whoisData = await getWhoisData(cleanDomain, timeout, debug, cache);
    if (!whoisData || whoisData.isAvailable) {
      log(`[whois] domain ${cleanDomain} is available or not registered`);
      return {
        domain: cleanDomain,
        isRegistered: false,
        isAvailable: true,
        status: [],
        registrar: null,
        nameServers: [],
        expirationDate: null,
        isExpired: false,
        daysUntilExpiration: null,
        isPendingDelete: false,
        isLocked: false
      };
    }
    const isRegistered = !!(whoisData.domainName || whoisData.creationDate || whoisData.registrar);
    let isExpired = false;
    let daysUntilExpiration = null;
    let expirationDate = null;
    if (whoisData.expirationDate) {
      expirationDate = new Date(whoisData.expirationDate);
      const now = /* @__PURE__ */ new Date();
      const expirationTime = expirationDate.getTime();
      const currentTime = now.getTime();
      isExpired = currentTime > expirationTime;
      if (!isExpired) {
        daysUntilExpiration = Math.floor((expirationTime - currentTime) / (1e3 * 60 * 60 * 24));
      }
      log(`[whois] domain ${cleanDomain} expires in ${daysUntilExpiration} days`);
    }
    const statusList = whoisData.status || [];
    const formattedStatusList = statusList.map((s) => {
      const statusCode = s.split(" ")[0];
      return statusCode;
    });
    const isPendingDelete = formattedStatusList.some((s) => s.toLowerCase().includes("pendingdelete") || s.toLowerCase().includes("redemption"));
    const isLocked = formattedStatusList.some((s) => s.toLowerCase().includes("clienttransferprohibited") || s.toLowerCase().includes("servertransferprohibited"));
    log(`[whois] domain ${cleanDomain} - registered: ${isRegistered}, expired: ${isExpired}, locked: ${isLocked}, pending delete: ${isPendingDelete}`);
    return {
      domain: cleanDomain,
      isRegistered,
      isAvailable: !isRegistered,
      status: formattedStatusList,
      registrar: whoisData.registrar || null,
      nameServers: whoisData.nameServers || [],
      expirationDate,
      isExpired,
      daysUntilExpiration,
      isPendingDelete,
      isLocked
    };
  } catch (ignoredError) {
    log(`[whois] error getting domain registration status for ${domain}: ${ignoredError instanceof Error ? ignoredError.message : "Unknown error"}`);
    return null;
  }
}

class RedisAdapter {
  constructor(redis, options = {}) {
    this.redis = redis;
    this.keyPrefix = options.keyPrefix || "email_validator:";
    this.defaultTtlMs = options.defaultTtlMs || 36e5;
    this.jsonSerializer = options.jsonSerializer || {
      stringify: (value) => {
        const processed = this.processDatesForSerialization(value);
        return JSON.stringify(processed);
      },
      parse: (value) => JSON.parse(value, (key, v) => {
        if (v && typeof v === "object" && v.__type === "Date") {
          return new Date(v.value);
        }
        return v;
      })
    };
  }
  getKey(key) {
    return `${this.keyPrefix}${key}`;
  }
  /**
   * Recursively process an object to convert Date instances to a serializable format
   */
  processDatesForSerialization(obj) {
    if (obj instanceof Date) {
      return { __type: "Date", value: obj.toISOString() };
    }
    if (obj && typeof obj === "object") {
      if (Array.isArray(obj)) {
        return obj.map((item) => this.processDatesForSerialization(item));
      }
      const result = {};
      for (const [key, value] of Object.entries(obj)) {
        result[key] = this.processDatesForSerialization(value);
      }
      return result;
    }
    return obj;
  }
  async get(key) {
    try {
      const value = await this.redis.get(this.getKey(key));
      if (value === null) {
        return null;
      }
      return this.jsonSerializer.parse(value);
    } catch (error) {
      console.error("Redis get error:", error);
      return null;
    }
  }
  async set(key, value, ttlMs) {
    try {
      const serializedValue = this.jsonSerializer.stringify(value);
      const ttl = ttlMs || this.defaultTtlMs;
      const ttlSeconds = Math.ceil(ttl / 1e3);
      await this.redis.set(this.getKey(key), serializedValue, "EX", ttlSeconds);
    } catch (error) {
      console.error("Redis set error:", error);
    }
  }
  async delete(key) {
    try {
      const result = await this.redis.del(this.getKey(key));
      return result > 0;
    } catch (error) {
      console.error("Redis delete error:", error);
      return false;
    }
  }
  async has(key) {
    try {
      const result = await this.redis.exists(this.getKey(key));
      return result > 0;
    } catch (error) {
      console.error("Redis exists error:", error);
      return false;
    }
  }
  async clear() {
    try {
      await this.redis.flushdb();
    } catch (error) {
      console.error("Redis clear error:", error);
    }
  }
  // size() is not applicable for Redis as it's a distributed store
  size() {
    return void 0;
  }
  /**
   * Helper method to delete only keys with the configured prefix
   * Requires Redis SCAN command which might not be available in all Redis clients
   */
  async clearPrefixed() {
    console.warn("clearPrefixed not implemented. Use clear() to flush the entire database.");
  }
}

async function verifyEmailBatch(params) {
  const { emailAddresses, concurrency = 5, timeout = 4e3, verifyMx = true, verifySmtp = false, checkDisposable = true, checkFree = true, detectName = false, nameDetectionMethod, suggestDomain = false, domainSuggestionMethod, commonDomains, skipMxForDisposable = false, skipDomainWhoisForDisposable = false, cache } = params;
  const startTime = Date.now();
  const results = /* @__PURE__ */ new Map();
  const batches = [];
  for (let i = 0; i < emailAddresses.length; i += concurrency) {
    batches.push(emailAddresses.slice(i, i + concurrency));
  }
  let totalValid = 0;
  let totalInvalid = 0;
  let totalErrors = 0;
  for (const batch of batches) {
    const batchPromises = batch.map(async (email) => {
      try {
        const result = await verifyEmail({
          emailAddress: email,
          timeout,
          verifyMx,
          verifySmtp,
          checkDisposable,
          checkFree,
          detectName,
          nameDetectionMethod,
          suggestDomain,
          domainSuggestionMethod,
          commonDomains,
          skipMxForDisposable,
          skipDomainWhoisForDisposable,
          cache
        });
        if (result.validFormat) {
          totalValid++;
        } else {
          totalInvalid++;
        }
        return { email, result };
      } catch (error) {
        totalErrors++;
        return {
          email,
          result: createErrorResult(email)
        };
      }
    });
    const batchResults = await Promise.all(batchPromises);
    for (const { email, result } of batchResults) {
      results.set(email, result);
    }
  }
  return {
    results,
    summary: {
      total: emailAddresses.length,
      valid: totalValid,
      invalid: totalInvalid,
      errors: totalErrors,
      processingTime: Date.now() - startTime
    }
  };
}
function createErrorResult(email, _error) {
  return {
    email,
    validFormat: false,
    validMx: null,
    validSmtp: null,
    isDisposable: false,
    isFree: false,
    metadata: {
      verificationTime: 0,
      cached: false,
      error: exports.VerificationErrorCode.smtpConnectionFailed
    }
  };
}

function isSpamEmail(email) {
  if (typeof email !== "string")
    return false;
  const atIndex = email.lastIndexOf("@");
  if (atIndex === -1)
    return false;
  const localPart = email.slice(0, atIndex);
  if (localPart.length < 16)
    return false;
  const cleaned = localPart.replace(/[._+-]/g, "");
  if (cleaned.length < 12)
    return false;
  if (!/^[A-Za-z0-9]+$/.test(cleaned))
    return false;
  const lettersOnly = cleaned.replace(/[0-9]/g, "");
  if (lettersOnly.length < 10)
    return false;
  function vowelRatio(str) {
    const vowels = (str.match(/[aeiouAEIOU]/g) || []).length;
    return vowels / str.length;
  }
  function uppercaseRatio(str) {
    const uppercase = (str.match(/[A-Z]/g) || []).length;
    return uppercase / str.length;
  }
  function hasRepeatedPatterns(str) {
    const patterns = /(.)\1{2,}/g;
    const matches = str.match(patterns);
    return matches !== null && matches.length > 0;
  }
  function hasConsonantClusters(str) {
    const clusters = str.match(/[^aeiouAEIOU]{3,}/g);
    return clusters !== null && clusters.length > 0;
  }
  function hasAlternatingCase(str) {
    if (str.length < 8)
      return false;
    let alternatingChanges = 0;
    for (let i = 1; i < str.length; i++) {
      const prevIsUpper = str[i - 1] >= "A" && str[i - 1] <= "Z";
      const currIsUpper = str[i] >= "A" && str[i] <= "Z";
      if (prevIsUpper !== currIsUpper) {
        alternatingChanges++;
      }
    }
    return alternatingChanges / str.length > 0.7;
  }
  const vowelR = vowelRatio(lettersOnly);
  const upperR = uppercaseRatio(lettersOnly);
  if (hasAlternatingCase(lettersOnly))
    return true;
  if (vowelR > 0.35)
    return false;
  if (upperR < 0.15)
    return false;
  if (hasRepeatedPatterns(localPart))
    return true;
  if (hasConsonantClusters(lettersOnly))
    return true;
  if (cleaned.length >= 20 && vowelR < 0.25)
    return true;
  return true;
}
function isSpamEmailLocalPart(email) {
  return isSpamEmail(email);
}

function isSpamName(name) {
  if (typeof name !== "string")
    return false;
  const cleaned = name.replace(/[,.;:!?]+$/, "").trim();
  const parts = cleaned.split(/\s+/);
  if (parts.length !== 2)
    return false;
  const [first, second] = parts;
  if (!/^[A-Za-z]+$/.test(first) || !/^[A-Za-z]+$/.test(second))
    return false;
  if (first.length < 16 || second.length < 16)
    return false;
  function vowelRatio(str) {
    const vowels = (str.match(/[aeiouAEIOU]/g) || []).length;
    return vowels / str.length;
  }
  const ratio1 = vowelRatio(first);
  const ratio2 = vowelRatio(second);
  const avgRatio = (ratio1 + ratio2) / 2;
  if (avgRatio > 0.35 && ratio1 > 0.25 && ratio2 > 0.25)
    return false;
  if (!/[A-Z]/.test(first) || !/[A-Z]/.test(second))
    return false;
  return true;
}

let disposableEmailProviders;
let freeEmailProviders;
async function isDisposableEmail(params) {
  const { emailOrDomain, cache, logger } = params;
  const log = logger || (() => {
  });
  const parts = emailOrDomain.split("@");
  const emailDomain = parts.length > 1 ? parts[1] : parts[0];
  if (!emailDomain) {
    return false;
  }
  const cacheStore = getCacheStore(cache, "disposable");
  let cached;
  try {
    cached = await cacheStore.get(emailDomain);
  } catch (ignoredError) {
    cached = null;
  }
  if (cached !== null && cached !== void 0) {
    log(`[isDisposableEmail] Cache hit for ${emailDomain}: ${cached.isDisposable}`);
    return cached.isDisposable;
  }
  if (!disposableEmailProviders) {
    disposableEmailProviders = new Set(require("./disposable-email-providers.json"));
  }
  const isDisposable = disposableEmailProviders.has(emailDomain);
  const richResult = {
    isDisposable,
    source: "disposable-email-providers.json",
    category: isDisposable ? "disposable" : void 0,
    checkedAt: Date.now()
  };
  try {
    await cacheStore.set(emailDomain, richResult);
    log(`[isDisposableEmail] Cached result for ${emailDomain}: ${isDisposable}`);
  } catch (ignoredError) {
    log(`[isDisposableEmail] Cache write error for ${emailDomain}`);
  }
  log(`[isDisposableEmail] Check result for ${emailDomain}: ${isDisposable}`);
  return isDisposable;
}
async function isFreeEmail(params) {
  const { emailOrDomain, cache, logger } = params;
  const log = logger || (() => {
  });
  const parts = emailOrDomain.split("@");
  const emailDomain = parts.length > 1 ? parts[1] : parts[0];
  if (!emailDomain) {
    return false;
  }
  const cacheStore = getCacheStore(cache, "free");
  let cached;
  try {
    cached = await cacheStore.get(emailDomain);
  } catch (ignoredError) {
    cached = null;
  }
  if (cached !== null && cached !== void 0) {
    log(`[isFreeEmail] Cache hit for ${emailDomain}: ${cached.isFree}`);
    return cached.isFree;
  }
  if (!freeEmailProviders) {
    freeEmailProviders = new Set(require("./free-email-providers.json"));
  }
  const isFree = freeEmailProviders.has(emailDomain);
  const richResult = {
    isFree,
    provider: isFree ? emailDomain : void 0,
    checkedAt: Date.now()
  };
  try {
    await cacheStore.set(emailDomain, richResult);
    log(`[isFreeEmail] Cached result for ${emailDomain}: ${isFree}`);
  } catch (ignoredError) {
    log(`[isFreeEmail] Cache write error for ${emailDomain}`);
  }
  log(`[isFreeEmail] Check result for ${emailDomain}: ${isFree}`);
  return isFree;
}
const domainPorts = {
  // 465 or 587
  // https://help.ovhcloud.com/csm/en-ca-web-paas-development-email?id=kb_article_view&sysparm_article=KB0053893
  "ovh.net": 465
};
async function verifyEmail(params) {
  var _a, _b;
  const { emailAddress, timeout = 4e3, verifyMx = true, verifySmtp = false, debug = false, checkDisposable = true, checkFree = true, detectName: detectName2 = false, nameDetectionMethod, suggestDomain: suggestDomain2 = true, domainSuggestionMethod, commonDomains, checkDomainAge = false, checkDomainRegistration = false, whoisTimeout = 5e3, skipMxForDisposable = false, skipDomainWhoisForDisposable = false } = params;
  const startTime = Date.now();
  const log = debug ? console.debug : (..._args) => {
  };
  const result = {
    email: emailAddress,
    validFormat: false,
    validMx: null,
    validSmtp: null,
    isDisposable: false,
    isFree: false,
    metadata: {
      verificationTime: 0,
      cached: false
    }
  };
  if (!isValidEmail(emailAddress)) {
    if (result.metadata) {
      result.metadata.verificationTime = Date.now() - startTime;
      result.metadata.error = exports.VerificationErrorCode.invalidFormat;
    }
    return result;
  }
  result.validFormat = true;
  if (detectName2) {
    result.detectedName = detectNameFromEmail({
      email: emailAddress,
      customMethod: nameDetectionMethod
    });
  }
  if (suggestDomain2) {
    const [, emailDomain] = emailAddress.split("@");
    if (emailDomain) {
      const suggestion = domainSuggestionMethod ? domainSuggestionMethod(emailDomain) : await suggestEmailDomain(emailAddress, commonDomains);
      if (suggestion) {
        result.domainSuggestion = suggestion;
      } else {
        result.domainSuggestion = null;
      }
    }
  }
  const [local, domain] = emailAddress.split("@");
  if (!domain || !local) {
    if (result.metadata) {
      result.metadata.verificationTime = Date.now() - startTime;
      result.metadata.error = exports.VerificationErrorCode.invalidFormat;
    }
    return result;
  }
  if (!await isValidEmailDomain(domain, params.cache)) {
    if (result.metadata) {
      result.metadata.verificationTime = Date.now() - startTime;
      result.metadata.error = exports.VerificationErrorCode.invalidDomain;
    }
    return result;
  }
  if (checkDisposable) {
    log(`[verifyEmail] Checking if ${emailAddress} is disposable email`);
    result.isDisposable = await isDisposableEmail({ emailOrDomain: emailAddress, cache: params.cache, logger: log });
    log(`[verifyEmail] Disposable check result: ${result.isDisposable}`);
    if (result.isDisposable && result.metadata) {
      result.metadata.error = exports.VerificationErrorCode.disposableEmail;
    }
  }
  if (checkFree) {
    log(`[verifyEmail] Checking if ${emailAddress} is free email provider`);
    result.isFree = await isFreeEmail({ emailOrDomain: emailAddress, cache: params.cache, logger: log });
    log(`[verifyEmail] Free email check result: ${result.isFree}`);
  }
  const shouldSkipMx = skipMxForDisposable && result.isDisposable;
  const shouldSkipDomainWhois = skipDomainWhoisForDisposable && result.isDisposable;
  if (shouldSkipMx) {
    log(`[verifyEmail] Skipping MX record check for disposable email: ${emailAddress}`);
  }
  if (shouldSkipDomainWhois) {
    log(`[verifyEmail] Skipping domain WHOIS checks for disposable email: ${emailAddress}`);
  }
  if (checkDomainAge && !shouldSkipDomainWhois) {
    log(`[verifyEmail] Checking domain age for ${domain}`);
    try {
      result.domainAge = await getDomainAge(domain, whoisTimeout, debug, params.cache);
      log(`[verifyEmail] Domain age result:`, result.domainAge ? `${result.domainAge.ageInDays} days` : "null");
    } catch (error) {
      log("[verifyEmail] Failed to get domain age", error);
      result.domainAge = null;
    }
  } else if (checkDomainAge && shouldSkipDomainWhois) {
    log(`[verifyEmail] Domain age check skipped due to disposable email and skipDomainWhoisForDisposable=true`);
  }
  if (checkDomainRegistration && !shouldSkipDomainWhois) {
    log(`[verifyEmail] Checking domain registration status for ${domain}`);
    try {
      result.domainRegistration = await getDomainRegistrationStatus(domain, whoisTimeout, debug, params.cache);
      log(`[verifyEmail] Domain registration result:`, ((_a = result.domainRegistration) === null || _a === void 0 ? void 0 : _a.isRegistered) ? "registered" : "not registered");
    } catch (error) {
      log("[verifyEmail] Failed to get domain registration status", error);
      result.domainRegistration = null;
    }
  } else if (checkDomainRegistration && shouldSkipDomainWhois) {
    log(`[verifyEmail] Domain registration check skipped due to disposable email and skipDomainWhoisForDisposable=true`);
  }
  if ((verifyMx || verifySmtp) && !shouldSkipMx) {
    log(`[verifyEmail] Checking MX records for ${domain}`);
    try {
      const mxRecords = await resolveMxRecords({ domain, cache: params.cache, logger: log });
      result.validMx = mxRecords.length > 0;
      result.mxRecords = mxRecords;
      log(`[verifyEmail] MX records found: ${mxRecords.length}, valid: ${result.validMx}`);
      if (!result.validMx && result.metadata) {
        result.metadata.error = exports.VerificationErrorCode.noMxRecords;
      }
      if (verifySmtp && mxRecords.length > 0) {
        const cacheKey = `${emailAddress}:smtp`;
        const smtpCacheInstance = getCacheStore(params.cache, "smtp");
        const cachedSmtp = await smtpCacheInstance.get(cacheKey);
        if (cachedSmtp !== null && cachedSmtp !== void 0) {
          result.canConnectSmtp = cachedSmtp.canConnectSmtp;
          result.hasFullInbox = cachedSmtp.hasFullInbox;
          result.isCatchAll = cachedSmtp.isCatchAll;
          result.isDeliverable = cachedSmtp.isDeliverable;
          result.isDisabled = cachedSmtp.isDisabled;
          result.validSmtp = (_b = cachedSmtp.isDeliverable) !== null && _b !== void 0 ? _b : null;
          log(`[verifyEmail] SMTP result from cache: ${result.validSmtp} for ${emailAddress}`);
          if (result.metadata) {
            result.metadata.cached = true;
          }
          if (detectName2 && !result.detectedName) {
            result.detectedName = detectNameFromEmail({
              email: emailAddress,
              customMethod: nameDetectionMethod
            });
          }
        } else {
          log(`[verifyEmail] Performing SMTP verification for ${emailAddress}`);
          let domainPort = params.smtpPort;
          if (!domainPort) {
            const mxDomain = psl.parse(mxRecords[0]);
            if ("domain" in mxDomain && mxDomain.domain) {
              domainPort = domainPorts[mxDomain.domain];
            }
          }
          const { smtpResult, cached, port } = await verifyMailboxSMTP({
            local,
            domain,
            mxRecords,
            options: {
              cache: params.cache,
              ports: domainPort ? [domainPort] : void 0,
              timeout,
              debug,
              maxRetries: params.retryAttempts
            }
          });
          await smtpCacheInstance.set(cacheKey, smtpResult);
          result.canConnectSmtp = smtpResult === null || smtpResult === void 0 ? void 0 : smtpResult.canConnectSmtp;
          result.hasFullInbox = smtpResult === null || smtpResult === void 0 ? void 0 : smtpResult.hasFullInbox;
          result.isCatchAll = smtpResult === null || smtpResult === void 0 ? void 0 : smtpResult.isCatchAll;
          result.isDeliverable = smtpResult === null || smtpResult === void 0 ? void 0 : smtpResult.isDeliverable;
          result.isDisabled = smtpResult === null || smtpResult === void 0 ? void 0 : smtpResult.isDisabled;
          if (!smtpResult.canConnectSmtp) {
            result.validSmtp = null;
          } else {
            result.validSmtp = smtpResult.isDeliverable;
          }
          if (result.metadata)
            result.metadata.cached = cached;
          log(`[verifyEmail] SMTP verification result: ${result.validSmtp} for ${emailAddress} (cached for future use)`);
        }
        if (result.validSmtp === false && result.metadata) {
          result.metadata.error = exports.VerificationErrorCode.mailboxNotFound;
        } else if (result.validSmtp === null && result.metadata) {
          result.metadata.error = exports.VerificationErrorCode.smtpConnectionFailed;
        }
      }
    } catch (error) {
      log("[verifyEmail] Failed to resolve MX records", error);
      result.validMx = false;
      result.mxRecords = null;
      if (result.metadata) {
        result.metadata.error = exports.VerificationErrorCode.noMxRecords;
      }
    }
  } else if ((verifyMx || verifySmtp) && shouldSkipMx) {
    log(`[verifyEmail] MX/SMTP checks skipped due to disposable email and skipMxForDisposable=true`);
  }
  if (result.metadata) {
    result.metadata.verificationTime = Date.now() - startTime;
  }
  return result;
}

exports.DEFAULT_CACHE_OPTIONS = DEFAULT_CACHE_OPTIONS;
exports.LRUAdapter = LRUAdapter;
exports.RedisAdapter = RedisAdapter;
exports.cleanNameForAlgorithm = cleanNameForAlgorithm;
exports.clearDefaultCache = clearDefaultCache;
exports.commonEmailDomains = commonEmailDomains;
exports.defaultDomainSuggestionMethod = defaultDomainSuggestionMethod;
exports.defaultNameDetectionMethod = defaultNameDetectionMethod;
exports.detectName = detectName;
exports.detectNameForAlgorithm = detectNameForAlgorithm;
exports.detectNameFromEmail = detectNameFromEmail;
exports.domainPorts = domainPorts;
exports.getCacheStore = getCacheStore;
exports.getDefaultCache = getDefaultCache;
exports.getDomainAge = getDomainAge;
exports.getDomainRegistrationStatus = getDomainRegistrationStatus;
exports.getDomainSimilarity = getDomainSimilarity;
exports.isCommonDomain = isCommonDomain;
exports.isDisposableEmail = isDisposableEmail;
exports.isFreeEmail = isFreeEmail;
exports.isSpamEmail = isSpamEmail;
exports.isSpamEmailLocalPart = isSpamEmailLocalPart;
exports.isSpamName = isSpamName;
exports.isValidEmail = isValidEmail;
exports.isValidEmailDomain = isValidEmailDomain;
exports.parseSmtpError = parseSmtpError;
exports.resetDefaultCache = resetDefaultCache;
exports.suggestDomain = suggestDomain;
exports.suggestEmailDomain = suggestEmailDomain;
exports.verifyEmail = verifyEmail;
exports.verifyEmailBatch = verifyEmailBatch;
